{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Dashboard"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/javascript": [
       "\n",
       "(function(root) {\n",
       "  function now() {\n",
       "    return new Date();\n",
       "  }\n",
       "\n",
       "  var force = true;\n",
       "\n",
       "  if (typeof root._bokeh_onload_callbacks === \"undefined\" || force === true) {\n",
       "    root._bokeh_onload_callbacks = [];\n",
       "    root._bokeh_is_loading = undefined;\n",
       "  }\n",
       "\n",
       "  if (typeof (root._bokeh_timeout) === \"undefined\" || force === true) {\n",
       "    root._bokeh_timeout = Date.now() + 5000;\n",
       "    root._bokeh_failed_load = false;\n",
       "  }\n",
       "\n",
       "  function run_callbacks() {\n",
       "    try {\n",
       "      root._bokeh_onload_callbacks.forEach(function(callback) {\n",
       "        if (callback != null)\n",
       "          callback();\n",
       "      });\n",
       "    } finally {\n",
       "      delete root._bokeh_onload_callbacks\n",
       "    }\n",
       "    console.debug(\"Bokeh: all callbacks have finished\");\n",
       "  }\n",
       "\n",
       "  function load_libs(css_urls, js_urls, callback) {\n",
       "    if (css_urls == null) css_urls = [];\n",
       "    if (js_urls == null) js_urls = [];\n",
       "\n",
       "    root._bokeh_onload_callbacks.push(callback);\n",
       "    if (root._bokeh_is_loading > 0) {\n",
       "      console.debug(\"Bokeh: BokehJS is being loaded, scheduling callback at\", now());\n",
       "      return null;\n",
       "    }\n",
       "    if (js_urls == null || js_urls.length === 0) {\n",
       "      run_callbacks();\n",
       "      return null;\n",
       "    }\n",
       "    console.debug(\"Bokeh: BokehJS not loaded, scheduling load and callback at\", now());\n",
       "    root._bokeh_is_loading = css_urls.length + js_urls.length;\n",
       "\n",
       "    function on_load() {\n",
       "      root._bokeh_is_loading--;\n",
       "      if (root._bokeh_is_loading === 0) {\n",
       "        console.debug(\"Bokeh: all BokehJS libraries/stylesheets loaded\");\n",
       "        run_callbacks()\n",
       "      }\n",
       "    }\n",
       "\n",
       "    function on_error() {\n",
       "      console.error(\"failed to load \" + url);\n",
       "    }\n",
       "\n",
       "    for (var i = 0; i < css_urls.length; i++) {\n",
       "      var url = css_urls[i];\n",
       "      const element = document.createElement(\"link\");\n",
       "      element.onload = on_load;\n",
       "      element.onerror = on_error;\n",
       "      element.rel = \"stylesheet\";\n",
       "      element.type = \"text/css\";\n",
       "      element.href = url;\n",
       "      console.debug(\"Bokeh: injecting link tag for BokehJS stylesheet: \", url);\n",
       "      document.body.appendChild(element);\n",
       "    }\n",
       "\n",
       "    var skip = [];\n",
       "    if (window.requirejs) {\n",
       "      window.requirejs.config({'paths': {'plotly': 'https://cdn.plot.ly/plotly-latest.min'}});\n",
       "      require([\"plotly\"], function(Plotly,) {\n",
       "        window.Plotly = Plotly;\n",
       "      })\n",
       "    }\n",
       "    if (((window['Plotly'] !== undefined) && (!(window['Plotly'] instanceof HTMLElement))) || window.requirejs) {\n",
       "      var urls = ['https://cdn.plot.ly/plotly-latest.min.js'];\n",
       "      for (var i = 0; i < urls.length; i++) {\n",
       "        skip.push(urls[i])\n",
       "      }\n",
       "    }\n",
       "    for (var i = 0; i < js_urls.length; i++) {\n",
       "      var url = js_urls[i];\n",
       "      if (skip.indexOf(url) >= 0) { on_load(); continue; }\n",
       "      var element = document.createElement('script');\n",
       "      element.onload = on_load;\n",
       "      element.onerror = on_error;\n",
       "      element.async = false;\n",
       "      element.src = url;\n",
       "      console.debug(\"Bokeh: injecting script tag for BokehJS library: \", url);\n",
       "      document.head.appendChild(element);\n",
       "    }\n",
       "\tif (!js_urls.length) {\n",
       "      on_load()\n",
       "    }\n",
       "  };\n",
       "\n",
       "  function inject_raw_css(css) {\n",
       "    const element = document.createElement(\"style\");\n",
       "    element.appendChild(document.createTextNode(css));\n",
       "    document.body.appendChild(element);\n",
       "  }\n",
       "\n",
       "  var js_urls = [\"https://code.jquery.com/jquery-3.4.1.min.js\", \"https://cdn.plot.ly/plotly-latest.min.js\", \"https://cdn.bokeh.org/bokeh/release/bokeh-2.2.2.min.js\", \"https://cdn.bokeh.org/bokeh/release/bokeh-widgets-2.2.2.min.js\", \"https://cdn.bokeh.org/bokeh/release/bokeh-tables-2.2.2.min.js\", \"https://unpkg.com/@holoviz/panel@^0.10.1/dist/panel.min.js\"];\n",
       "  var css_urls = [\"https://unpkg.com/@holoviz/panel@0.10.1/dist/css/alerts.css\", \"https://unpkg.com/@holoviz/panel@0.10.1/dist/css/card.css\", \"https://unpkg.com/@holoviz/panel@0.10.1/dist/css/dataframe.css\", \"https://unpkg.com/@holoviz/panel@0.10.1/dist/css/json.css\", \"https://unpkg.com/@holoviz/panel@0.10.1/dist/css/markdown.css\", \"https://unpkg.com/@holoviz/panel@0.10.1/dist/css/widgets.css\"];\n",
       "\n",
       "  var inline_js = [\n",
       "    function(Bokeh) {\n",
       "      Bokeh.set_log_level(\"info\");\n",
       "    },\n",
       "    function(Bokeh) {} // ensure no trailing comma for IE\n",
       "  ];\n",
       "\n",
       "  function run_inline_js() {\n",
       "    if ((root.Bokeh !== undefined) || (force === true)) {\n",
       "      for (var i = 0; i < inline_js.length; i++) {\n",
       "        inline_js[i].call(root, root.Bokeh);\n",
       "      }} else if (Date.now() < root._bokeh_timeout) {\n",
       "      setTimeout(run_inline_js, 100);\n",
       "    } else if (!root._bokeh_failed_load) {\n",
       "      console.log(\"Bokeh: BokehJS failed to load within specified timeout.\");\n",
       "      root._bokeh_failed_load = true;\n",
       "    }\n",
       "  }\n",
       "\n",
       "  if (root._bokeh_is_loading === 0) {\n",
       "    console.debug(\"Bokeh: BokehJS loaded, going straight to plotting\");\n",
       "    run_inline_js();\n",
       "  } else {\n",
       "    load_libs(css_urls, js_urls, function() {\n",
       "      console.debug(\"Bokeh: BokehJS plotting callback run at\", now());\n",
       "      run_inline_js();\n",
       "    });\n",
       "  }\n",
       "}(window));"
      ],
      "application/vnd.holoviews_load.v0+json": "\n(function(root) {\n  function now() {\n    return new Date();\n  }\n\n  var force = true;\n\n  if (typeof root._bokeh_onload_callbacks === \"undefined\" || force === true) {\n    root._bokeh_onload_callbacks = [];\n    root._bokeh_is_loading = undefined;\n  }\n\n  if (typeof (root._bokeh_timeout) === \"undefined\" || force === true) {\n    root._bokeh_timeout = Date.now() + 5000;\n    root._bokeh_failed_load = false;\n  }\n\n  function run_callbacks() {\n    try {\n      root._bokeh_onload_callbacks.forEach(function(callback) {\n        if (callback != null)\n          callback();\n      });\n    } finally {\n      delete root._bokeh_onload_callbacks\n    }\n    console.debug(\"Bokeh: all callbacks have finished\");\n  }\n\n  function load_libs(css_urls, js_urls, callback) {\n    if (css_urls == null) css_urls = [];\n    if (js_urls == null) js_urls = [];\n\n    root._bokeh_onload_callbacks.push(callback);\n    if (root._bokeh_is_loading > 0) {\n      console.debug(\"Bokeh: BokehJS is being loaded, scheduling callback at\", now());\n      return null;\n    }\n    if (js_urls == null || js_urls.length === 0) {\n      run_callbacks();\n      return null;\n    }\n    console.debug(\"Bokeh: BokehJS not loaded, scheduling load and callback at\", now());\n    root._bokeh_is_loading = css_urls.length + js_urls.length;\n\n    function on_load() {\n      root._bokeh_is_loading--;\n      if (root._bokeh_is_loading === 0) {\n        console.debug(\"Bokeh: all BokehJS libraries/stylesheets loaded\");\n        run_callbacks()\n      }\n    }\n\n    function on_error() {\n      console.error(\"failed to load \" + url);\n    }\n\n    for (var i = 0; i < css_urls.length; i++) {\n      var url = css_urls[i];\n      const element = document.createElement(\"link\");\n      element.onload = on_load;\n      element.onerror = on_error;\n      element.rel = \"stylesheet\";\n      element.type = \"text/css\";\n      element.href = url;\n      console.debug(\"Bokeh: injecting link tag for BokehJS stylesheet: \", url);\n      document.body.appendChild(element);\n    }\n\n    var skip = [];\n    if (window.requirejs) {\n      window.requirejs.config({'paths': {'plotly': 'https://cdn.plot.ly/plotly-latest.min'}});\n      require([\"plotly\"], function(Plotly,) {\n        window.Plotly = Plotly;\n      })\n    }\n    if (((window['Plotly'] !== undefined) && (!(window['Plotly'] instanceof HTMLElement))) || window.requirejs) {\n      var urls = ['https://cdn.plot.ly/plotly-latest.min.js'];\n      for (var i = 0; i < urls.length; i++) {\n        skip.push(urls[i])\n      }\n    }\n    for (var i = 0; i < js_urls.length; i++) {\n      var url = js_urls[i];\n      if (skip.indexOf(url) >= 0) { on_load(); continue; }\n      var element = document.createElement('script');\n      element.onload = on_load;\n      element.onerror = on_error;\n      element.async = false;\n      element.src = url;\n      console.debug(\"Bokeh: injecting script tag for BokehJS library: \", url);\n      document.head.appendChild(element);\n    }\n\tif (!js_urls.length) {\n      on_load()\n    }\n  };\n\n  function inject_raw_css(css) {\n    const element = document.createElement(\"style\");\n    element.appendChild(document.createTextNode(css));\n    document.body.appendChild(element);\n  }\n\n  var js_urls = [\"https://code.jquery.com/jquery-3.4.1.min.js\", \"https://cdn.plot.ly/plotly-latest.min.js\", \"https://cdn.bokeh.org/bokeh/release/bokeh-2.2.2.min.js\", \"https://cdn.bokeh.org/bokeh/release/bokeh-widgets-2.2.2.min.js\", \"https://cdn.bokeh.org/bokeh/release/bokeh-tables-2.2.2.min.js\", \"https://unpkg.com/@holoviz/panel@^0.10.1/dist/panel.min.js\"];\n  var css_urls = [\"https://unpkg.com/@holoviz/panel@0.10.1/dist/css/alerts.css\", \"https://unpkg.com/@holoviz/panel@0.10.1/dist/css/card.css\", \"https://unpkg.com/@holoviz/panel@0.10.1/dist/css/dataframe.css\", \"https://unpkg.com/@holoviz/panel@0.10.1/dist/css/json.css\", \"https://unpkg.com/@holoviz/panel@0.10.1/dist/css/markdown.css\", \"https://unpkg.com/@holoviz/panel@0.10.1/dist/css/widgets.css\"];\n\n  var inline_js = [\n    function(Bokeh) {\n      Bokeh.set_log_level(\"info\");\n    },\n    function(Bokeh) {} // ensure no trailing comma for IE\n  ];\n\n  function run_inline_js() {\n    if ((root.Bokeh !== undefined) || (force === true)) {\n      for (var i = 0; i < inline_js.length; i++) {\n        inline_js[i].call(root, root.Bokeh);\n      }} else if (Date.now() < root._bokeh_timeout) {\n      setTimeout(run_inline_js, 100);\n    } else if (!root._bokeh_failed_load) {\n      console.log(\"Bokeh: BokehJS failed to load within specified timeout.\");\n      root._bokeh_failed_load = true;\n    }\n  }\n\n  if (root._bokeh_is_loading === 0) {\n    console.debug(\"Bokeh: BokehJS loaded, going straight to plotting\");\n    run_inline_js();\n  } else {\n    load_libs(css_urls, js_urls, function() {\n      console.debug(\"Bokeh: BokehJS plotting callback run at\", now());\n      run_inline_js();\n    });\n  }\n}(window));"
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "application/javascript": [
       "\n",
       "if ((window.PyViz === undefined) || (window.PyViz instanceof HTMLElement)) {\n",
       "  window.PyViz = {comms: {}, comm_status:{}, kernels:{}, receivers: {}, plot_index: []}\n",
       "}\n",
       "\n",
       "\n",
       "    function JupyterCommManager() {\n",
       "    }\n",
       "\n",
       "    JupyterCommManager.prototype.register_target = function(plot_id, comm_id, msg_handler) {\n",
       "      if (window.comm_manager || ((window.Jupyter !== undefined) && (Jupyter.notebook.kernel != null))) {\n",
       "        var comm_manager = window.comm_manager || Jupyter.notebook.kernel.comm_manager;\n",
       "        comm_manager.register_target(comm_id, function(comm) {\n",
       "          comm.on_msg(msg_handler);\n",
       "        });\n",
       "      } else if ((plot_id in window.PyViz.kernels) && (window.PyViz.kernels[plot_id])) {\n",
       "        window.PyViz.kernels[plot_id].registerCommTarget(comm_id, function(comm) {\n",
       "          comm.onMsg = msg_handler;\n",
       "        });\n",
       "      } else if (typeof google != 'undefined' && google.colab.kernel != null) {\n",
       "        google.colab.kernel.comms.registerTarget(comm_id, (comm) => {\n",
       "          var messages = comm.messages[Symbol.asyncIterator]();\n",
       "          function processIteratorResult(result) {\n",
       "            var message = result.value;\n",
       "            console.log(message)\n",
       "            var content = {data: message.data, comm_id};\n",
       "            var buffers = []\n",
       "            for (var buffer of message.buffers || []) {\n",
       "              buffers.push(new DataView(buffer))\n",
       "            }\n",
       "            var metadata = message.metadata || {};\n",
       "            var msg = {content, buffers, metadata}\n",
       "            msg_handler(msg);\n",
       "            return messages.next().then(processIteratorResult);\n",
       "          }\n",
       "          return messages.next().then(processIteratorResult);\n",
       "        })\n",
       "      }\n",
       "    }\n",
       "\n",
       "    JupyterCommManager.prototype.get_client_comm = function(plot_id, comm_id, msg_handler) {\n",
       "      if (comm_id in window.PyViz.comms) {\n",
       "        return window.PyViz.comms[comm_id];\n",
       "      } else if (window.comm_manager || ((window.Jupyter !== undefined) && (Jupyter.notebook.kernel != null))) {\n",
       "        var comm_manager = window.comm_manager || Jupyter.notebook.kernel.comm_manager;\n",
       "        var comm = comm_manager.new_comm(comm_id, {}, {}, {}, comm_id);\n",
       "        if (msg_handler) {\n",
       "          comm.on_msg(msg_handler);\n",
       "        }\n",
       "      } else if ((plot_id in window.PyViz.kernels) && (window.PyViz.kernels[plot_id])) {\n",
       "        var comm = window.PyViz.kernels[plot_id].connectToComm(comm_id);\n",
       "        comm.open();\n",
       "        if (msg_handler) {\n",
       "          comm.onMsg = msg_handler;\n",
       "        }\n",
       "      } else if (typeof google != 'undefined' && google.colab.kernel != null) {\n",
       "        var comm_promise = google.colab.kernel.comms.open(comm_id)\n",
       "        comm_promise.then((comm) => {\n",
       "          window.PyViz.comms[comm_id] = comm;\n",
       "          if (msg_handler) {\n",
       "            var messages = comm.messages[Symbol.asyncIterator]();\n",
       "            function processIteratorResult(result) {\n",
       "              var message = result.value;\n",
       "              var content = {data: message.data};\n",
       "              var metadata = message.metadata || {comm_id};\n",
       "              var msg = {content, metadata}\n",
       "              msg_handler(msg);\n",
       "              return messages.next().then(processIteratorResult);\n",
       "            }\n",
       "            return messages.next().then(processIteratorResult);\n",
       "          }\n",
       "        }) \n",
       "        var sendClosure = (data, metadata, buffers, disposeOnDone) => {\n",
       "          return comm_promise.then((comm) => {\n",
       "            comm.send(data, metadata, buffers, disposeOnDone);\n",
       "          });\n",
       "        };\n",
       "        var comm = {\n",
       "          send: sendClosure\n",
       "        };\n",
       "      }\n",
       "      window.PyViz.comms[comm_id] = comm;\n",
       "      return comm;\n",
       "    }\n",
       "    window.PyViz.comm_manager = new JupyterCommManager();\n",
       "    \n",
       "\n",
       "\n",
       "var JS_MIME_TYPE = 'application/javascript';\n",
       "var HTML_MIME_TYPE = 'text/html';\n",
       "var EXEC_MIME_TYPE = 'application/vnd.holoviews_exec.v0+json';\n",
       "var CLASS_NAME = 'output';\n",
       "\n",
       "/**\n",
       " * Render data to the DOM node\n",
       " */\n",
       "function render(props, node) {\n",
       "  var div = document.createElement(\"div\");\n",
       "  var script = document.createElement(\"script\");\n",
       "  node.appendChild(div);\n",
       "  node.appendChild(script);\n",
       "}\n",
       "\n",
       "/**\n",
       " * Handle when a new output is added\n",
       " */\n",
       "function handle_add_output(event, handle) {\n",
       "  var output_area = handle.output_area;\n",
       "  var output = handle.output;\n",
       "  if ((output.data == undefined) || (!output.data.hasOwnProperty(EXEC_MIME_TYPE))) {\n",
       "    return\n",
       "  }\n",
       "  var id = output.metadata[EXEC_MIME_TYPE][\"id\"];\n",
       "  var toinsert = output_area.element.find(\".\" + CLASS_NAME.split(' ')[0]);\n",
       "  if (id !== undefined) {\n",
       "    var nchildren = toinsert.length;\n",
       "    var html_node = toinsert[nchildren-1].children[0];\n",
       "    html_node.innerHTML = output.data[HTML_MIME_TYPE];\n",
       "    var scripts = [];\n",
       "    var nodelist = html_node.querySelectorAll(\"script\");\n",
       "    for (var i in nodelist) {\n",
       "      if (nodelist.hasOwnProperty(i)) {\n",
       "        scripts.push(nodelist[i])\n",
       "      }\n",
       "    }\n",
       "\n",
       "    scripts.forEach( function (oldScript) {\n",
       "      var newScript = document.createElement(\"script\");\n",
       "      var attrs = [];\n",
       "      var nodemap = oldScript.attributes;\n",
       "      for (var j in nodemap) {\n",
       "        if (nodemap.hasOwnProperty(j)) {\n",
       "          attrs.push(nodemap[j])\n",
       "        }\n",
       "      }\n",
       "      attrs.forEach(function(attr) { newScript.setAttribute(attr.name, attr.value) });\n",
       "      newScript.appendChild(document.createTextNode(oldScript.innerHTML));\n",
       "      oldScript.parentNode.replaceChild(newScript, oldScript);\n",
       "    });\n",
       "    if (JS_MIME_TYPE in output.data) {\n",
       "      toinsert[nchildren-1].children[1].textContent = output.data[JS_MIME_TYPE];\n",
       "    }\n",
       "    output_area._hv_plot_id = id;\n",
       "    if ((window.Bokeh !== undefined) && (id in Bokeh.index)) {\n",
       "      window.PyViz.plot_index[id] = Bokeh.index[id];\n",
       "    } else {\n",
       "      window.PyViz.plot_index[id] = null;\n",
       "    }\n",
       "  } else if (output.metadata[EXEC_MIME_TYPE][\"server_id\"] !== undefined) {\n",
       "    var bk_div = document.createElement(\"div\");\n",
       "    bk_div.innerHTML = output.data[HTML_MIME_TYPE];\n",
       "    var script_attrs = bk_div.children[0].attributes;\n",
       "    for (var i = 0; i < script_attrs.length; i++) {\n",
       "      toinsert[toinsert.length - 1].childNodes[1].setAttribute(script_attrs[i].name, script_attrs[i].value);\n",
       "    }\n",
       "    // store reference to server id on output_area\n",
       "    output_area._bokeh_server_id = output.metadata[EXEC_MIME_TYPE][\"server_id\"];\n",
       "  }\n",
       "}\n",
       "\n",
       "/**\n",
       " * Handle when an output is cleared or removed\n",
       " */\n",
       "function handle_clear_output(event, handle) {\n",
       "  var id = handle.cell.output_area._hv_plot_id;\n",
       "  var server_id = handle.cell.output_area._bokeh_server_id;\n",
       "  if (((id === undefined) || !(id in PyViz.plot_index)) && (server_id !== undefined)) { return; }\n",
       "  var comm = window.PyViz.comm_manager.get_client_comm(\"hv-extension-comm\", \"hv-extension-comm\", function () {});\n",
       "  if (server_id !== null) {\n",
       "    comm.send({event_type: 'server_delete', 'id': server_id});\n",
       "    return;\n",
       "  } else if (comm !== null) {\n",
       "    comm.send({event_type: 'delete', 'id': id});\n",
       "  }\n",
       "  delete PyViz.plot_index[id];\n",
       "  if ((window.Bokeh !== undefined) & (id in window.Bokeh.index)) {\n",
       "    var doc = window.Bokeh.index[id].model.document\n",
       "    doc.clear();\n",
       "    const i = window.Bokeh.documents.indexOf(doc);\n",
       "    if (i > -1) {\n",
       "      window.Bokeh.documents.splice(i, 1);\n",
       "    }\n",
       "  }\n",
       "}\n",
       "\n",
       "/**\n",
       " * Handle kernel restart event\n",
       " */\n",
       "function handle_kernel_cleanup(event, handle) {\n",
       "  delete PyViz.comms[\"hv-extension-comm\"];\n",
       "  window.PyViz.plot_index = {}\n",
       "}\n",
       "\n",
       "/**\n",
       " * Handle update_display_data messages\n",
       " */\n",
       "function handle_update_output(event, handle) {\n",
       "  handle_clear_output(event, {cell: {output_area: handle.output_area}})\n",
       "  handle_add_output(event, handle)\n",
       "}\n",
       "\n",
       "function register_renderer(events, OutputArea) {\n",
       "  function append_mime(data, metadata, element) {\n",
       "    // create a DOM node to render to\n",
       "    var toinsert = this.create_output_subarea(\n",
       "    metadata,\n",
       "    CLASS_NAME,\n",
       "    EXEC_MIME_TYPE\n",
       "    );\n",
       "    this.keyboard_manager.register_events(toinsert);\n",
       "    // Render to node\n",
       "    var props = {data: data, metadata: metadata[EXEC_MIME_TYPE]};\n",
       "    render(props, toinsert[0]);\n",
       "    element.append(toinsert);\n",
       "    return toinsert\n",
       "  }\n",
       "\n",
       "  events.on('output_added.OutputArea', handle_add_output);\n",
       "  events.on('output_updated.OutputArea', handle_update_output);\n",
       "  events.on('clear_output.CodeCell', handle_clear_output);\n",
       "  events.on('delete.Cell', handle_clear_output);\n",
       "  events.on('kernel_ready.Kernel', handle_kernel_cleanup);\n",
       "\n",
       "  OutputArea.prototype.register_mime_type(EXEC_MIME_TYPE, append_mime, {\n",
       "    safe: true,\n",
       "    index: 0\n",
       "  });\n",
       "}\n",
       "\n",
       "if (window.Jupyter !== undefined) {\n",
       "  try {\n",
       "    var events = require('base/js/events');\n",
       "    var OutputArea = require('notebook/js/outputarea').OutputArea;\n",
       "    if (OutputArea.prototype.mime_types().indexOf(EXEC_MIME_TYPE) == -1) {\n",
       "      register_renderer(events, OutputArea);\n",
       "    }\n",
       "  } catch(err) {\n",
       "  }\n",
       "}\n"
      ],
      "application/vnd.holoviews_load.v0+json": "\nif ((window.PyViz === undefined) || (window.PyViz instanceof HTMLElement)) {\n  window.PyViz = {comms: {}, comm_status:{}, kernels:{}, receivers: {}, plot_index: []}\n}\n\n\n    function JupyterCommManager() {\n    }\n\n    JupyterCommManager.prototype.register_target = function(plot_id, comm_id, msg_handler) {\n      if (window.comm_manager || ((window.Jupyter !== undefined) && (Jupyter.notebook.kernel != null))) {\n        var comm_manager = window.comm_manager || Jupyter.notebook.kernel.comm_manager;\n        comm_manager.register_target(comm_id, function(comm) {\n          comm.on_msg(msg_handler);\n        });\n      } else if ((plot_id in window.PyViz.kernels) && (window.PyViz.kernels[plot_id])) {\n        window.PyViz.kernels[plot_id].registerCommTarget(comm_id, function(comm) {\n          comm.onMsg = msg_handler;\n        });\n      } else if (typeof google != 'undefined' && google.colab.kernel != null) {\n        google.colab.kernel.comms.registerTarget(comm_id, (comm) => {\n          var messages = comm.messages[Symbol.asyncIterator]();\n          function processIteratorResult(result) {\n            var message = result.value;\n            console.log(message)\n            var content = {data: message.data, comm_id};\n            var buffers = []\n            for (var buffer of message.buffers || []) {\n              buffers.push(new DataView(buffer))\n            }\n            var metadata = message.metadata || {};\n            var msg = {content, buffers, metadata}\n            msg_handler(msg);\n            return messages.next().then(processIteratorResult);\n          }\n          return messages.next().then(processIteratorResult);\n        })\n      }\n    }\n\n    JupyterCommManager.prototype.get_client_comm = function(plot_id, comm_id, msg_handler) {\n      if (comm_id in window.PyViz.comms) {\n        return window.PyViz.comms[comm_id];\n      } else if (window.comm_manager || ((window.Jupyter !== undefined) && (Jupyter.notebook.kernel != null))) {\n        var comm_manager = window.comm_manager || Jupyter.notebook.kernel.comm_manager;\n        var comm = comm_manager.new_comm(comm_id, {}, {}, {}, comm_id);\n        if (msg_handler) {\n          comm.on_msg(msg_handler);\n        }\n      } else if ((plot_id in window.PyViz.kernels) && (window.PyViz.kernels[plot_id])) {\n        var comm = window.PyViz.kernels[plot_id].connectToComm(comm_id);\n        comm.open();\n        if (msg_handler) {\n          comm.onMsg = msg_handler;\n        }\n      } else if (typeof google != 'undefined' && google.colab.kernel != null) {\n        var comm_promise = google.colab.kernel.comms.open(comm_id)\n        comm_promise.then((comm) => {\n          window.PyViz.comms[comm_id] = comm;\n          if (msg_handler) {\n            var messages = comm.messages[Symbol.asyncIterator]();\n            function processIteratorResult(result) {\n              var message = result.value;\n              var content = {data: message.data};\n              var metadata = message.metadata || {comm_id};\n              var msg = {content, metadata}\n              msg_handler(msg);\n              return messages.next().then(processIteratorResult);\n            }\n            return messages.next().then(processIteratorResult);\n          }\n        }) \n        var sendClosure = (data, metadata, buffers, disposeOnDone) => {\n          return comm_promise.then((comm) => {\n            comm.send(data, metadata, buffers, disposeOnDone);\n          });\n        };\n        var comm = {\n          send: sendClosure\n        };\n      }\n      window.PyViz.comms[comm_id] = comm;\n      return comm;\n    }\n    window.PyViz.comm_manager = new JupyterCommManager();\n    \n\n\nvar JS_MIME_TYPE = 'application/javascript';\nvar HTML_MIME_TYPE = 'text/html';\nvar EXEC_MIME_TYPE = 'application/vnd.holoviews_exec.v0+json';\nvar CLASS_NAME = 'output';\n\n/**\n * Render data to the DOM node\n */\nfunction render(props, node) {\n  var div = document.createElement(\"div\");\n  var script = document.createElement(\"script\");\n  node.appendChild(div);\n  node.appendChild(script);\n}\n\n/**\n * Handle when a new output is added\n */\nfunction handle_add_output(event, handle) {\n  var output_area = handle.output_area;\n  var output = handle.output;\n  if ((output.data == undefined) || (!output.data.hasOwnProperty(EXEC_MIME_TYPE))) {\n    return\n  }\n  var id = output.metadata[EXEC_MIME_TYPE][\"id\"];\n  var toinsert = output_area.element.find(\".\" + CLASS_NAME.split(' ')[0]);\n  if (id !== undefined) {\n    var nchildren = toinsert.length;\n    var html_node = toinsert[nchildren-1].children[0];\n    html_node.innerHTML = output.data[HTML_MIME_TYPE];\n    var scripts = [];\n    var nodelist = html_node.querySelectorAll(\"script\");\n    for (var i in nodelist) {\n      if (nodelist.hasOwnProperty(i)) {\n        scripts.push(nodelist[i])\n      }\n    }\n\n    scripts.forEach( function (oldScript) {\n      var newScript = document.createElement(\"script\");\n      var attrs = [];\n      var nodemap = oldScript.attributes;\n      for (var j in nodemap) {\n        if (nodemap.hasOwnProperty(j)) {\n          attrs.push(nodemap[j])\n        }\n      }\n      attrs.forEach(function(attr) { newScript.setAttribute(attr.name, attr.value) });\n      newScript.appendChild(document.createTextNode(oldScript.innerHTML));\n      oldScript.parentNode.replaceChild(newScript, oldScript);\n    });\n    if (JS_MIME_TYPE in output.data) {\n      toinsert[nchildren-1].children[1].textContent = output.data[JS_MIME_TYPE];\n    }\n    output_area._hv_plot_id = id;\n    if ((window.Bokeh !== undefined) && (id in Bokeh.index)) {\n      window.PyViz.plot_index[id] = Bokeh.index[id];\n    } else {\n      window.PyViz.plot_index[id] = null;\n    }\n  } else if (output.metadata[EXEC_MIME_TYPE][\"server_id\"] !== undefined) {\n    var bk_div = document.createElement(\"div\");\n    bk_div.innerHTML = output.data[HTML_MIME_TYPE];\n    var script_attrs = bk_div.children[0].attributes;\n    for (var i = 0; i < script_attrs.length; i++) {\n      toinsert[toinsert.length - 1].childNodes[1].setAttribute(script_attrs[i].name, script_attrs[i].value);\n    }\n    // store reference to server id on output_area\n    output_area._bokeh_server_id = output.metadata[EXEC_MIME_TYPE][\"server_id\"];\n  }\n}\n\n/**\n * Handle when an output is cleared or removed\n */\nfunction handle_clear_output(event, handle) {\n  var id = handle.cell.output_area._hv_plot_id;\n  var server_id = handle.cell.output_area._bokeh_server_id;\n  if (((id === undefined) || !(id in PyViz.plot_index)) && (server_id !== undefined)) { return; }\n  var comm = window.PyViz.comm_manager.get_client_comm(\"hv-extension-comm\", \"hv-extension-comm\", function () {});\n  if (server_id !== null) {\n    comm.send({event_type: 'server_delete', 'id': server_id});\n    return;\n  } else if (comm !== null) {\n    comm.send({event_type: 'delete', 'id': id});\n  }\n  delete PyViz.plot_index[id];\n  if ((window.Bokeh !== undefined) & (id in window.Bokeh.index)) {\n    var doc = window.Bokeh.index[id].model.document\n    doc.clear();\n    const i = window.Bokeh.documents.indexOf(doc);\n    if (i > -1) {\n      window.Bokeh.documents.splice(i, 1);\n    }\n  }\n}\n\n/**\n * Handle kernel restart event\n */\nfunction handle_kernel_cleanup(event, handle) {\n  delete PyViz.comms[\"hv-extension-comm\"];\n  window.PyViz.plot_index = {}\n}\n\n/**\n * Handle update_display_data messages\n */\nfunction handle_update_output(event, handle) {\n  handle_clear_output(event, {cell: {output_area: handle.output_area}})\n  handle_add_output(event, handle)\n}\n\nfunction register_renderer(events, OutputArea) {\n  function append_mime(data, metadata, element) {\n    // create a DOM node to render to\n    var toinsert = this.create_output_subarea(\n    metadata,\n    CLASS_NAME,\n    EXEC_MIME_TYPE\n    );\n    this.keyboard_manager.register_events(toinsert);\n    // Render to node\n    var props = {data: data, metadata: metadata[EXEC_MIME_TYPE]};\n    render(props, toinsert[0]);\n    element.append(toinsert);\n    return toinsert\n  }\n\n  events.on('output_added.OutputArea', handle_add_output);\n  events.on('output_updated.OutputArea', handle_update_output);\n  events.on('clear_output.CodeCell', handle_clear_output);\n  events.on('delete.Cell', handle_clear_output);\n  events.on('kernel_ready.Kernel', handle_kernel_cleanup);\n\n  OutputArea.prototype.register_mime_type(EXEC_MIME_TYPE, append_mime, {\n    safe: true,\n    index: 0\n  });\n}\n\nif (window.Jupyter !== undefined) {\n  try {\n    var events = require('base/js/events');\n    var OutputArea = require('notebook/js/outputarea').OutputArea;\n    if (OutputArea.prototype.mime_types().indexOf(EXEC_MIME_TYPE) == -1) {\n      register_renderer(events, OutputArea);\n    }\n  } catch(err) {\n  }\n}\n"
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# export\n",
    "import os\n",
    "import re\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import panel as pn\n",
    "import webbrowser\n",
    "pn.extension('plotly')\n",
    "\n",
    "import pyteomics.fasta\n",
    "from io import StringIO\n",
    "import plotly.graph_objects as go\n",
    "\n",
    "from pepmap.importing import import_data\n",
    "from pepmap.preprocessing import format_input_data\n",
    "from pepmap.sequenceplot import plot_peptide_traces\n",
    "from pepmap.uniprot_integration import uniprot_feature_dict\n",
    "from pepmap.sequenceplot import uniprot_color_dict"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "path_to_folder_fasta_files = 'data'\n",
    "full_fasta = None\n",
    "full_uniprot = None\n",
    "ac_gene_conversion = None"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Style of the dashboard"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/javascript": [
       "\n",
       "(function(root) {\n",
       "  function now() {\n",
       "    return new Date();\n",
       "  }\n",
       "\n",
       "  var force = true;\n",
       "\n",
       "  if (typeof root._bokeh_onload_callbacks === \"undefined\" || force === true) {\n",
       "    root._bokeh_onload_callbacks = [];\n",
       "    root._bokeh_is_loading = undefined;\n",
       "  }\n",
       "\n",
       "  if (typeof (root._bokeh_timeout) === \"undefined\" || force === true) {\n",
       "    root._bokeh_timeout = Date.now() + 5000;\n",
       "    root._bokeh_failed_load = false;\n",
       "  }\n",
       "\n",
       "  function run_callbacks() {\n",
       "    try {\n",
       "      root._bokeh_onload_callbacks.forEach(function(callback) {\n",
       "        if (callback != null)\n",
       "          callback();\n",
       "      });\n",
       "    } finally {\n",
       "      delete root._bokeh_onload_callbacks\n",
       "    }\n",
       "    console.debug(\"Bokeh: all callbacks have finished\");\n",
       "  }\n",
       "\n",
       "  function load_libs(css_urls, js_urls, callback) {\n",
       "    if (css_urls == null) css_urls = [];\n",
       "    if (js_urls == null) js_urls = [];\n",
       "\n",
       "    root._bokeh_onload_callbacks.push(callback);\n",
       "    if (root._bokeh_is_loading > 0) {\n",
       "      console.debug(\"Bokeh: BokehJS is being loaded, scheduling callback at\", now());\n",
       "      return null;\n",
       "    }\n",
       "    if (js_urls == null || js_urls.length === 0) {\n",
       "      run_callbacks();\n",
       "      return null;\n",
       "    }\n",
       "    console.debug(\"Bokeh: BokehJS not loaded, scheduling load and callback at\", now());\n",
       "    root._bokeh_is_loading = css_urls.length + js_urls.length;\n",
       "\n",
       "    function on_load() {\n",
       "      root._bokeh_is_loading--;\n",
       "      if (root._bokeh_is_loading === 0) {\n",
       "        console.debug(\"Bokeh: all BokehJS libraries/stylesheets loaded\");\n",
       "        run_callbacks()\n",
       "      }\n",
       "    }\n",
       "\n",
       "    function on_error() {\n",
       "      console.error(\"failed to load \" + url);\n",
       "    }\n",
       "\n",
       "    for (var i = 0; i < css_urls.length; i++) {\n",
       "      var url = css_urls[i];\n",
       "      const element = document.createElement(\"link\");\n",
       "      element.onload = on_load;\n",
       "      element.onerror = on_error;\n",
       "      element.rel = \"stylesheet\";\n",
       "      element.type = \"text/css\";\n",
       "      element.href = url;\n",
       "      console.debug(\"Bokeh: injecting link tag for BokehJS stylesheet: \", url);\n",
       "      document.body.appendChild(element);\n",
       "    }\n",
       "\n",
       "    var skip = [];\n",
       "    if (window.requirejs) {\n",
       "      window.requirejs.config({'paths': {'plotly': 'https://cdn.plot.ly/plotly-latest.min'}});\n",
       "      require([], function() {\n",
       "      })\n",
       "    }\n",
       "    if (((window['Plotly'] !== undefined) && (!(window['Plotly'] instanceof HTMLElement))) || window.requirejs) {\n",
       "      var urls = ['https://cdn.plot.ly/plotly-latest.min.js'];\n",
       "      for (var i = 0; i < urls.length; i++) {\n",
       "        skip.push(urls[i])\n",
       "      }\n",
       "    }\n",
       "    for (var i = 0; i < js_urls.length; i++) {\n",
       "      var url = js_urls[i];\n",
       "      if (skip.indexOf(url) >= 0) { on_load(); continue; }\n",
       "      var element = document.createElement('script');\n",
       "      element.onload = on_load;\n",
       "      element.onerror = on_error;\n",
       "      element.async = false;\n",
       "      element.src = url;\n",
       "      console.debug(\"Bokeh: injecting script tag for BokehJS library: \", url);\n",
       "      document.head.appendChild(element);\n",
       "    }\n",
       "\tif (!js_urls.length) {\n",
       "      on_load()\n",
       "    }\n",
       "  };\n",
       "\n",
       "  function inject_raw_css(css) {\n",
       "    const element = document.createElement(\"style\");\n",
       "    element.appendChild(document.createTextNode(css));\n",
       "    document.body.appendChild(element);\n",
       "  }\n",
       "\n",
       "  var js_urls = [\"https://code.jquery.com/jquery-3.4.1.min.js\", \"https://cdn.plot.ly/plotly-latest.min.js\", \"https://cdn.bokeh.org/bokeh/release/bokeh-2.2.2.min.js\", \"https://cdn.bokeh.org/bokeh/release/bokeh-widgets-2.2.2.min.js\", \"https://cdn.bokeh.org/bokeh/release/bokeh-tables-2.2.2.min.js\", \"https://unpkg.com/@holoviz/panel@^0.10.1/dist/panel.min.js\"];\n",
       "  var css_urls = [\"https://unpkg.com/@holoviz/panel@0.10.1/dist/css/alerts.css\", \"https://unpkg.com/@holoviz/panel@0.10.1/dist/css/card.css\", \"https://unpkg.com/@holoviz/panel@0.10.1/dist/css/dataframe.css\", \"https://unpkg.com/@holoviz/panel@0.10.1/dist/css/json.css\", \"https://unpkg.com/@holoviz/panel@0.10.1/dist/css/markdown.css\", \"https://unpkg.com/@holoviz/panel@0.10.1/dist/css/widgets.css\"];\n",
       "\n",
       "  var inline_js = [\n",
       "    function(Bokeh) {\n",
       "      inject_raw_css(\"\\n\\n.bk-root .bk-btn-default.bk-active {\\n    background-color: #AECCE6;\\n}\\n\\n.bk.accordion {\\n    font-size: smaller;\\n}\\n\\n.bk-root .bk-btn {\\n    padding: 6px 0px 6px 0px;\\n}\\n\\n.bk.uniprot_options {\\n    margin: 17px 6px 12px 6px;\\n    background-color: white;\\n    font-size: 13px;\\n    font-weight: initial;\\n    text-align: center;\\n    height: 55px;\\n    border: 1px solid #e0e0e0;\\n    border-radius: 2px;\\n}\\n\\n.bk-root .bk-btn-primary {\\n    background-color: #045082;\\n    font-size: 11px;\\n    font-weight: 700;\\n    text-transform: uppercase;\\n    letter-spacing: 1.5px;\\n}\\n\\n.bk.card-title {\\n    font-size: 13px;\\n    font-weight: initial;\\n}\\n\\nh1 {\\n    color: #045082;\\n    font-size: 45px;\\n    line-height: 0.6;\\n    text-align: center;\\n}\\n\\nh4 {\\n    color: #173245;\\n    font-size: 18px;\\n    line-height: 0.6;\\n    text-align: center;\\n}\\n\\n.codehilite {\\n    color: #173245;\\n    font-size: 14px;\\n    background-color: #eaeaea;\\n}\\n\\n.codehilite .s2 {\\n    color: #173245;\\n    font-weight: bold;\\n}\\n\\n.codehilite .ow {\\n    color: #045082;\\n    font-weight: initial;\\n}\\n\\n.bk.opt {\\n    position: relative;\\n    display: block;\\n    left: 75px;\\n    top: 0px;\\n    width: 80px;\\n    height: 80px;\\n}\\n\\n.bk.main-part {\\n    background-color: #EAEAEA;\\n    font-size: 17px;\\n    line-height: 23px;\\n    letter-spacing: 0px;\\n    font-weight: 500;\\n    color: #045082;\\n    text-align: center;\\n    position: relative !important;\\n    margin-left: auto;\\n    margin-right: auto;\\n    width: 40%;\\n}\\n\\n.bk.selection_box {\\n    background-color: #EAEAEA;\\n}\\n\\n.bk.alert-danger {\\n    background-color: #EAEAEA;\\n    color: #c72d3b;\\n    border: 0px #EAEAEA solid;\\n    padding: 0;\\n    font-size: 12px;\\n}\\n\\n\");\n",
       "    },\n",
       "    function(Bokeh) {\n",
       "      Bokeh.set_log_level(\"info\");\n",
       "    },\n",
       "    function(Bokeh) {} // ensure no trailing comma for IE\n",
       "  ];\n",
       "\n",
       "  function run_inline_js() {\n",
       "    if ((root.Bokeh !== undefined) || (force === true)) {\n",
       "      for (var i = 0; i < inline_js.length; i++) {\n",
       "        inline_js[i].call(root, root.Bokeh);\n",
       "      }} else if (Date.now() < root._bokeh_timeout) {\n",
       "      setTimeout(run_inline_js, 100);\n",
       "    } else if (!root._bokeh_failed_load) {\n",
       "      console.log(\"Bokeh: BokehJS failed to load within specified timeout.\");\n",
       "      root._bokeh_failed_load = true;\n",
       "    }\n",
       "  }\n",
       "\n",
       "  if (root._bokeh_is_loading === 0) {\n",
       "    console.debug(\"Bokeh: BokehJS loaded, going straight to plotting\");\n",
       "    run_inline_js();\n",
       "  } else {\n",
       "    load_libs(css_urls, js_urls, function() {\n",
       "      console.debug(\"Bokeh: BokehJS plotting callback run at\", now());\n",
       "      run_inline_js();\n",
       "    });\n",
       "  }\n",
       "}(window));"
      ],
      "application/vnd.holoviews_load.v0+json": "\n(function(root) {\n  function now() {\n    return new Date();\n  }\n\n  var force = true;\n\n  if (typeof root._bokeh_onload_callbacks === \"undefined\" || force === true) {\n    root._bokeh_onload_callbacks = [];\n    root._bokeh_is_loading = undefined;\n  }\n\n  if (typeof (root._bokeh_timeout) === \"undefined\" || force === true) {\n    root._bokeh_timeout = Date.now() + 5000;\n    root._bokeh_failed_load = false;\n  }\n\n  function run_callbacks() {\n    try {\n      root._bokeh_onload_callbacks.forEach(function(callback) {\n        if (callback != null)\n          callback();\n      });\n    } finally {\n      delete root._bokeh_onload_callbacks\n    }\n    console.debug(\"Bokeh: all callbacks have finished\");\n  }\n\n  function load_libs(css_urls, js_urls, callback) {\n    if (css_urls == null) css_urls = [];\n    if (js_urls == null) js_urls = [];\n\n    root._bokeh_onload_callbacks.push(callback);\n    if (root._bokeh_is_loading > 0) {\n      console.debug(\"Bokeh: BokehJS is being loaded, scheduling callback at\", now());\n      return null;\n    }\n    if (js_urls == null || js_urls.length === 0) {\n      run_callbacks();\n      return null;\n    }\n    console.debug(\"Bokeh: BokehJS not loaded, scheduling load and callback at\", now());\n    root._bokeh_is_loading = css_urls.length + js_urls.length;\n\n    function on_load() {\n      root._bokeh_is_loading--;\n      if (root._bokeh_is_loading === 0) {\n        console.debug(\"Bokeh: all BokehJS libraries/stylesheets loaded\");\n        run_callbacks()\n      }\n    }\n\n    function on_error() {\n      console.error(\"failed to load \" + url);\n    }\n\n    for (var i = 0; i < css_urls.length; i++) {\n      var url = css_urls[i];\n      const element = document.createElement(\"link\");\n      element.onload = on_load;\n      element.onerror = on_error;\n      element.rel = \"stylesheet\";\n      element.type = \"text/css\";\n      element.href = url;\n      console.debug(\"Bokeh: injecting link tag for BokehJS stylesheet: \", url);\n      document.body.appendChild(element);\n    }\n\n    var skip = [];\n    if (window.requirejs) {\n      window.requirejs.config({'paths': {'plotly': 'https://cdn.plot.ly/plotly-latest.min'}});\n      require([], function() {\n      })\n    }\n    if (((window['Plotly'] !== undefined) && (!(window['Plotly'] instanceof HTMLElement))) || window.requirejs) {\n      var urls = ['https://cdn.plot.ly/plotly-latest.min.js'];\n      for (var i = 0; i < urls.length; i++) {\n        skip.push(urls[i])\n      }\n    }\n    for (var i = 0; i < js_urls.length; i++) {\n      var url = js_urls[i];\n      if (skip.indexOf(url) >= 0) { on_load(); continue; }\n      var element = document.createElement('script');\n      element.onload = on_load;\n      element.onerror = on_error;\n      element.async = false;\n      element.src = url;\n      console.debug(\"Bokeh: injecting script tag for BokehJS library: \", url);\n      document.head.appendChild(element);\n    }\n\tif (!js_urls.length) {\n      on_load()\n    }\n  };\n\n  function inject_raw_css(css) {\n    const element = document.createElement(\"style\");\n    element.appendChild(document.createTextNode(css));\n    document.body.appendChild(element);\n  }\n\n  var js_urls = [\"https://code.jquery.com/jquery-3.4.1.min.js\", \"https://cdn.plot.ly/plotly-latest.min.js\", \"https://cdn.bokeh.org/bokeh/release/bokeh-2.2.2.min.js\", \"https://cdn.bokeh.org/bokeh/release/bokeh-widgets-2.2.2.min.js\", \"https://cdn.bokeh.org/bokeh/release/bokeh-tables-2.2.2.min.js\", \"https://unpkg.com/@holoviz/panel@^0.10.1/dist/panel.min.js\"];\n  var css_urls = [\"https://unpkg.com/@holoviz/panel@0.10.1/dist/css/alerts.css\", \"https://unpkg.com/@holoviz/panel@0.10.1/dist/css/card.css\", \"https://unpkg.com/@holoviz/panel@0.10.1/dist/css/dataframe.css\", \"https://unpkg.com/@holoviz/panel@0.10.1/dist/css/json.css\", \"https://unpkg.com/@holoviz/panel@0.10.1/dist/css/markdown.css\", \"https://unpkg.com/@holoviz/panel@0.10.1/dist/css/widgets.css\"];\n\n  var inline_js = [\n    function(Bokeh) {\n      inject_raw_css(\"\\n\\n.bk-root .bk-btn-default.bk-active {\\n    background-color: #AECCE6;\\n}\\n\\n.bk.accordion {\\n    font-size: smaller;\\n}\\n\\n.bk-root .bk-btn {\\n    padding: 6px 0px 6px 0px;\\n}\\n\\n.bk.uniprot_options {\\n    margin: 17px 6px 12px 6px;\\n    background-color: white;\\n    font-size: 13px;\\n    font-weight: initial;\\n    text-align: center;\\n    height: 55px;\\n    border: 1px solid #e0e0e0;\\n    border-radius: 2px;\\n}\\n\\n.bk-root .bk-btn-primary {\\n    background-color: #045082;\\n    font-size: 11px;\\n    font-weight: 700;\\n    text-transform: uppercase;\\n    letter-spacing: 1.5px;\\n}\\n\\n.bk.card-title {\\n    font-size: 13px;\\n    font-weight: initial;\\n}\\n\\nh1 {\\n    color: #045082;\\n    font-size: 45px;\\n    line-height: 0.6;\\n    text-align: center;\\n}\\n\\nh4 {\\n    color: #173245;\\n    font-size: 18px;\\n    line-height: 0.6;\\n    text-align: center;\\n}\\n\\n.codehilite {\\n    color: #173245;\\n    font-size: 14px;\\n    background-color: #eaeaea;\\n}\\n\\n.codehilite .s2 {\\n    color: #173245;\\n    font-weight: bold;\\n}\\n\\n.codehilite .ow {\\n    color: #045082;\\n    font-weight: initial;\\n}\\n\\n.bk.opt {\\n    position: relative;\\n    display: block;\\n    left: 75px;\\n    top: 0px;\\n    width: 80px;\\n    height: 80px;\\n}\\n\\n.bk.main-part {\\n    background-color: #EAEAEA;\\n    font-size: 17px;\\n    line-height: 23px;\\n    letter-spacing: 0px;\\n    font-weight: 500;\\n    color: #045082;\\n    text-align: center;\\n    position: relative !important;\\n    margin-left: auto;\\n    margin-right: auto;\\n    width: 40%;\\n}\\n\\n.bk.selection_box {\\n    background-color: #EAEAEA;\\n}\\n\\n.bk.alert-danger {\\n    background-color: #EAEAEA;\\n    color: #c72d3b;\\n    border: 0px #EAEAEA solid;\\n    padding: 0;\\n    font-size: 12px;\\n}\\n\\n\");\n    },\n    function(Bokeh) {\n      Bokeh.set_log_level(\"info\");\n    },\n    function(Bokeh) {} // ensure no trailing comma for IE\n  ];\n\n  function run_inline_js() {\n    if ((root.Bokeh !== undefined) || (force === true)) {\n      for (var i = 0; i < inline_js.length; i++) {\n        inline_js[i].call(root, root.Bokeh);\n      }} else if (Date.now() < root._bokeh_timeout) {\n      setTimeout(run_inline_js, 100);\n    } else if (!root._bokeh_failed_load) {\n      console.log(\"Bokeh: BokehJS failed to load within specified timeout.\");\n      root._bokeh_failed_load = true;\n    }\n  }\n\n  if (root._bokeh_is_loading === 0) {\n    console.debug(\"Bokeh: BokehJS loaded, going straight to plotting\");\n    run_inline_js();\n  } else {\n    load_libs(css_urls, js_urls, function() {\n      console.debug(\"Bokeh: BokehJS plotting callback run at\", now());\n      run_inline_js();\n    });\n  }\n}(window));"
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "application/javascript": [
       "\n",
       "if ((window.PyViz === undefined) || (window.PyViz instanceof HTMLElement)) {\n",
       "  window.PyViz = {comms: {}, comm_status:{}, kernels:{}, receivers: {}, plot_index: []}\n",
       "}\n",
       "\n",
       "\n",
       "    function JupyterCommManager() {\n",
       "    }\n",
       "\n",
       "    JupyterCommManager.prototype.register_target = function(plot_id, comm_id, msg_handler) {\n",
       "      if (window.comm_manager || ((window.Jupyter !== undefined) && (Jupyter.notebook.kernel != null))) {\n",
       "        var comm_manager = window.comm_manager || Jupyter.notebook.kernel.comm_manager;\n",
       "        comm_manager.register_target(comm_id, function(comm) {\n",
       "          comm.on_msg(msg_handler);\n",
       "        });\n",
       "      } else if ((plot_id in window.PyViz.kernels) && (window.PyViz.kernels[plot_id])) {\n",
       "        window.PyViz.kernels[plot_id].registerCommTarget(comm_id, function(comm) {\n",
       "          comm.onMsg = msg_handler;\n",
       "        });\n",
       "      } else if (typeof google != 'undefined' && google.colab.kernel != null) {\n",
       "        google.colab.kernel.comms.registerTarget(comm_id, (comm) => {\n",
       "          var messages = comm.messages[Symbol.asyncIterator]();\n",
       "          function processIteratorResult(result) {\n",
       "            var message = result.value;\n",
       "            console.log(message)\n",
       "            var content = {data: message.data, comm_id};\n",
       "            var buffers = []\n",
       "            for (var buffer of message.buffers || []) {\n",
       "              buffers.push(new DataView(buffer))\n",
       "            }\n",
       "            var metadata = message.metadata || {};\n",
       "            var msg = {content, buffers, metadata}\n",
       "            msg_handler(msg);\n",
       "            return messages.next().then(processIteratorResult);\n",
       "          }\n",
       "          return messages.next().then(processIteratorResult);\n",
       "        })\n",
       "      }\n",
       "    }\n",
       "\n",
       "    JupyterCommManager.prototype.get_client_comm = function(plot_id, comm_id, msg_handler) {\n",
       "      if (comm_id in window.PyViz.comms) {\n",
       "        return window.PyViz.comms[comm_id];\n",
       "      } else if (window.comm_manager || ((window.Jupyter !== undefined) && (Jupyter.notebook.kernel != null))) {\n",
       "        var comm_manager = window.comm_manager || Jupyter.notebook.kernel.comm_manager;\n",
       "        var comm = comm_manager.new_comm(comm_id, {}, {}, {}, comm_id);\n",
       "        if (msg_handler) {\n",
       "          comm.on_msg(msg_handler);\n",
       "        }\n",
       "      } else if ((plot_id in window.PyViz.kernels) && (window.PyViz.kernels[plot_id])) {\n",
       "        var comm = window.PyViz.kernels[plot_id].connectToComm(comm_id);\n",
       "        comm.open();\n",
       "        if (msg_handler) {\n",
       "          comm.onMsg = msg_handler;\n",
       "        }\n",
       "      } else if (typeof google != 'undefined' && google.colab.kernel != null) {\n",
       "        var comm_promise = google.colab.kernel.comms.open(comm_id)\n",
       "        comm_promise.then((comm) => {\n",
       "          window.PyViz.comms[comm_id] = comm;\n",
       "          if (msg_handler) {\n",
       "            var messages = comm.messages[Symbol.asyncIterator]();\n",
       "            function processIteratorResult(result) {\n",
       "              var message = result.value;\n",
       "              var content = {data: message.data};\n",
       "              var metadata = message.metadata || {comm_id};\n",
       "              var msg = {content, metadata}\n",
       "              msg_handler(msg);\n",
       "              return messages.next().then(processIteratorResult);\n",
       "            }\n",
       "            return messages.next().then(processIteratorResult);\n",
       "          }\n",
       "        }) \n",
       "        var sendClosure = (data, metadata, buffers, disposeOnDone) => {\n",
       "          return comm_promise.then((comm) => {\n",
       "            comm.send(data, metadata, buffers, disposeOnDone);\n",
       "          });\n",
       "        };\n",
       "        var comm = {\n",
       "          send: sendClosure\n",
       "        };\n",
       "      }\n",
       "      window.PyViz.comms[comm_id] = comm;\n",
       "      return comm;\n",
       "    }\n",
       "    window.PyViz.comm_manager = new JupyterCommManager();\n",
       "    \n",
       "\n",
       "\n",
       "var JS_MIME_TYPE = 'application/javascript';\n",
       "var HTML_MIME_TYPE = 'text/html';\n",
       "var EXEC_MIME_TYPE = 'application/vnd.holoviews_exec.v0+json';\n",
       "var CLASS_NAME = 'output';\n",
       "\n",
       "/**\n",
       " * Render data to the DOM node\n",
       " */\n",
       "function render(props, node) {\n",
       "  var div = document.createElement(\"div\");\n",
       "  var script = document.createElement(\"script\");\n",
       "  node.appendChild(div);\n",
       "  node.appendChild(script);\n",
       "}\n",
       "\n",
       "/**\n",
       " * Handle when a new output is added\n",
       " */\n",
       "function handle_add_output(event, handle) {\n",
       "  var output_area = handle.output_area;\n",
       "  var output = handle.output;\n",
       "  if ((output.data == undefined) || (!output.data.hasOwnProperty(EXEC_MIME_TYPE))) {\n",
       "    return\n",
       "  }\n",
       "  var id = output.metadata[EXEC_MIME_TYPE][\"id\"];\n",
       "  var toinsert = output_area.element.find(\".\" + CLASS_NAME.split(' ')[0]);\n",
       "  if (id !== undefined) {\n",
       "    var nchildren = toinsert.length;\n",
       "    var html_node = toinsert[nchildren-1].children[0];\n",
       "    html_node.innerHTML = output.data[HTML_MIME_TYPE];\n",
       "    var scripts = [];\n",
       "    var nodelist = html_node.querySelectorAll(\"script\");\n",
       "    for (var i in nodelist) {\n",
       "      if (nodelist.hasOwnProperty(i)) {\n",
       "        scripts.push(nodelist[i])\n",
       "      }\n",
       "    }\n",
       "\n",
       "    scripts.forEach( function (oldScript) {\n",
       "      var newScript = document.createElement(\"script\");\n",
       "      var attrs = [];\n",
       "      var nodemap = oldScript.attributes;\n",
       "      for (var j in nodemap) {\n",
       "        if (nodemap.hasOwnProperty(j)) {\n",
       "          attrs.push(nodemap[j])\n",
       "        }\n",
       "      }\n",
       "      attrs.forEach(function(attr) { newScript.setAttribute(attr.name, attr.value) });\n",
       "      newScript.appendChild(document.createTextNode(oldScript.innerHTML));\n",
       "      oldScript.parentNode.replaceChild(newScript, oldScript);\n",
       "    });\n",
       "    if (JS_MIME_TYPE in output.data) {\n",
       "      toinsert[nchildren-1].children[1].textContent = output.data[JS_MIME_TYPE];\n",
       "    }\n",
       "    output_area._hv_plot_id = id;\n",
       "    if ((window.Bokeh !== undefined) && (id in Bokeh.index)) {\n",
       "      window.PyViz.plot_index[id] = Bokeh.index[id];\n",
       "    } else {\n",
       "      window.PyViz.plot_index[id] = null;\n",
       "    }\n",
       "  } else if (output.metadata[EXEC_MIME_TYPE][\"server_id\"] !== undefined) {\n",
       "    var bk_div = document.createElement(\"div\");\n",
       "    bk_div.innerHTML = output.data[HTML_MIME_TYPE];\n",
       "    var script_attrs = bk_div.children[0].attributes;\n",
       "    for (var i = 0; i < script_attrs.length; i++) {\n",
       "      toinsert[toinsert.length - 1].childNodes[1].setAttribute(script_attrs[i].name, script_attrs[i].value);\n",
       "    }\n",
       "    // store reference to server id on output_area\n",
       "    output_area._bokeh_server_id = output.metadata[EXEC_MIME_TYPE][\"server_id\"];\n",
       "  }\n",
       "}\n",
       "\n",
       "/**\n",
       " * Handle when an output is cleared or removed\n",
       " */\n",
       "function handle_clear_output(event, handle) {\n",
       "  var id = handle.cell.output_area._hv_plot_id;\n",
       "  var server_id = handle.cell.output_area._bokeh_server_id;\n",
       "  if (((id === undefined) || !(id in PyViz.plot_index)) && (server_id !== undefined)) { return; }\n",
       "  var comm = window.PyViz.comm_manager.get_client_comm(\"hv-extension-comm\", \"hv-extension-comm\", function () {});\n",
       "  if (server_id !== null) {\n",
       "    comm.send({event_type: 'server_delete', 'id': server_id});\n",
       "    return;\n",
       "  } else if (comm !== null) {\n",
       "    comm.send({event_type: 'delete', 'id': id});\n",
       "  }\n",
       "  delete PyViz.plot_index[id];\n",
       "  if ((window.Bokeh !== undefined) & (id in window.Bokeh.index)) {\n",
       "    var doc = window.Bokeh.index[id].model.document\n",
       "    doc.clear();\n",
       "    const i = window.Bokeh.documents.indexOf(doc);\n",
       "    if (i > -1) {\n",
       "      window.Bokeh.documents.splice(i, 1);\n",
       "    }\n",
       "  }\n",
       "}\n",
       "\n",
       "/**\n",
       " * Handle kernel restart event\n",
       " */\n",
       "function handle_kernel_cleanup(event, handle) {\n",
       "  delete PyViz.comms[\"hv-extension-comm\"];\n",
       "  window.PyViz.plot_index = {}\n",
       "}\n",
       "\n",
       "/**\n",
       " * Handle update_display_data messages\n",
       " */\n",
       "function handle_update_output(event, handle) {\n",
       "  handle_clear_output(event, {cell: {output_area: handle.output_area}})\n",
       "  handle_add_output(event, handle)\n",
       "}\n",
       "\n",
       "function register_renderer(events, OutputArea) {\n",
       "  function append_mime(data, metadata, element) {\n",
       "    // create a DOM node to render to\n",
       "    var toinsert = this.create_output_subarea(\n",
       "    metadata,\n",
       "    CLASS_NAME,\n",
       "    EXEC_MIME_TYPE\n",
       "    );\n",
       "    this.keyboard_manager.register_events(toinsert);\n",
       "    // Render to node\n",
       "    var props = {data: data, metadata: metadata[EXEC_MIME_TYPE]};\n",
       "    render(props, toinsert[0]);\n",
       "    element.append(toinsert);\n",
       "    return toinsert\n",
       "  }\n",
       "\n",
       "  events.on('output_added.OutputArea', handle_add_output);\n",
       "  events.on('output_updated.OutputArea', handle_update_output);\n",
       "  events.on('clear_output.CodeCell', handle_clear_output);\n",
       "  events.on('delete.Cell', handle_clear_output);\n",
       "  events.on('kernel_ready.Kernel', handle_kernel_cleanup);\n",
       "\n",
       "  OutputArea.prototype.register_mime_type(EXEC_MIME_TYPE, append_mime, {\n",
       "    safe: true,\n",
       "    index: 0\n",
       "  });\n",
       "}\n",
       "\n",
       "if (window.Jupyter !== undefined) {\n",
       "  try {\n",
       "    var events = require('base/js/events');\n",
       "    var OutputArea = require('notebook/js/outputarea').OutputArea;\n",
       "    if (OutputArea.prototype.mime_types().indexOf(EXEC_MIME_TYPE) == -1) {\n",
       "      register_renderer(events, OutputArea);\n",
       "    }\n",
       "  } catch(err) {\n",
       "  }\n",
       "}\n"
      ],
      "application/vnd.holoviews_load.v0+json": "\nif ((window.PyViz === undefined) || (window.PyViz instanceof HTMLElement)) {\n  window.PyViz = {comms: {}, comm_status:{}, kernels:{}, receivers: {}, plot_index: []}\n}\n\n\n    function JupyterCommManager() {\n    }\n\n    JupyterCommManager.prototype.register_target = function(plot_id, comm_id, msg_handler) {\n      if (window.comm_manager || ((window.Jupyter !== undefined) && (Jupyter.notebook.kernel != null))) {\n        var comm_manager = window.comm_manager || Jupyter.notebook.kernel.comm_manager;\n        comm_manager.register_target(comm_id, function(comm) {\n          comm.on_msg(msg_handler);\n        });\n      } else if ((plot_id in window.PyViz.kernels) && (window.PyViz.kernels[plot_id])) {\n        window.PyViz.kernels[plot_id].registerCommTarget(comm_id, function(comm) {\n          comm.onMsg = msg_handler;\n        });\n      } else if (typeof google != 'undefined' && google.colab.kernel != null) {\n        google.colab.kernel.comms.registerTarget(comm_id, (comm) => {\n          var messages = comm.messages[Symbol.asyncIterator]();\n          function processIteratorResult(result) {\n            var message = result.value;\n            console.log(message)\n            var content = {data: message.data, comm_id};\n            var buffers = []\n            for (var buffer of message.buffers || []) {\n              buffers.push(new DataView(buffer))\n            }\n            var metadata = message.metadata || {};\n            var msg = {content, buffers, metadata}\n            msg_handler(msg);\n            return messages.next().then(processIteratorResult);\n          }\n          return messages.next().then(processIteratorResult);\n        })\n      }\n    }\n\n    JupyterCommManager.prototype.get_client_comm = function(plot_id, comm_id, msg_handler) {\n      if (comm_id in window.PyViz.comms) {\n        return window.PyViz.comms[comm_id];\n      } else if (window.comm_manager || ((window.Jupyter !== undefined) && (Jupyter.notebook.kernel != null))) {\n        var comm_manager = window.comm_manager || Jupyter.notebook.kernel.comm_manager;\n        var comm = comm_manager.new_comm(comm_id, {}, {}, {}, comm_id);\n        if (msg_handler) {\n          comm.on_msg(msg_handler);\n        }\n      } else if ((plot_id in window.PyViz.kernels) && (window.PyViz.kernels[plot_id])) {\n        var comm = window.PyViz.kernels[plot_id].connectToComm(comm_id);\n        comm.open();\n        if (msg_handler) {\n          comm.onMsg = msg_handler;\n        }\n      } else if (typeof google != 'undefined' && google.colab.kernel != null) {\n        var comm_promise = google.colab.kernel.comms.open(comm_id)\n        comm_promise.then((comm) => {\n          window.PyViz.comms[comm_id] = comm;\n          if (msg_handler) {\n            var messages = comm.messages[Symbol.asyncIterator]();\n            function processIteratorResult(result) {\n              var message = result.value;\n              var content = {data: message.data};\n              var metadata = message.metadata || {comm_id};\n              var msg = {content, metadata}\n              msg_handler(msg);\n              return messages.next().then(processIteratorResult);\n            }\n            return messages.next().then(processIteratorResult);\n          }\n        }) \n        var sendClosure = (data, metadata, buffers, disposeOnDone) => {\n          return comm_promise.then((comm) => {\n            comm.send(data, metadata, buffers, disposeOnDone);\n          });\n        };\n        var comm = {\n          send: sendClosure\n        };\n      }\n      window.PyViz.comms[comm_id] = comm;\n      return comm;\n    }\n    window.PyViz.comm_manager = new JupyterCommManager();\n    \n\n\nvar JS_MIME_TYPE = 'application/javascript';\nvar HTML_MIME_TYPE = 'text/html';\nvar EXEC_MIME_TYPE = 'application/vnd.holoviews_exec.v0+json';\nvar CLASS_NAME = 'output';\n\n/**\n * Render data to the DOM node\n */\nfunction render(props, node) {\n  var div = document.createElement(\"div\");\n  var script = document.createElement(\"script\");\n  node.appendChild(div);\n  node.appendChild(script);\n}\n\n/**\n * Handle when a new output is added\n */\nfunction handle_add_output(event, handle) {\n  var output_area = handle.output_area;\n  var output = handle.output;\n  if ((output.data == undefined) || (!output.data.hasOwnProperty(EXEC_MIME_TYPE))) {\n    return\n  }\n  var id = output.metadata[EXEC_MIME_TYPE][\"id\"];\n  var toinsert = output_area.element.find(\".\" + CLASS_NAME.split(' ')[0]);\n  if (id !== undefined) {\n    var nchildren = toinsert.length;\n    var html_node = toinsert[nchildren-1].children[0];\n    html_node.innerHTML = output.data[HTML_MIME_TYPE];\n    var scripts = [];\n    var nodelist = html_node.querySelectorAll(\"script\");\n    for (var i in nodelist) {\n      if (nodelist.hasOwnProperty(i)) {\n        scripts.push(nodelist[i])\n      }\n    }\n\n    scripts.forEach( function (oldScript) {\n      var newScript = document.createElement(\"script\");\n      var attrs = [];\n      var nodemap = oldScript.attributes;\n      for (var j in nodemap) {\n        if (nodemap.hasOwnProperty(j)) {\n          attrs.push(nodemap[j])\n        }\n      }\n      attrs.forEach(function(attr) { newScript.setAttribute(attr.name, attr.value) });\n      newScript.appendChild(document.createTextNode(oldScript.innerHTML));\n      oldScript.parentNode.replaceChild(newScript, oldScript);\n    });\n    if (JS_MIME_TYPE in output.data) {\n      toinsert[nchildren-1].children[1].textContent = output.data[JS_MIME_TYPE];\n    }\n    output_area._hv_plot_id = id;\n    if ((window.Bokeh !== undefined) && (id in Bokeh.index)) {\n      window.PyViz.plot_index[id] = Bokeh.index[id];\n    } else {\n      window.PyViz.plot_index[id] = null;\n    }\n  } else if (output.metadata[EXEC_MIME_TYPE][\"server_id\"] !== undefined) {\n    var bk_div = document.createElement(\"div\");\n    bk_div.innerHTML = output.data[HTML_MIME_TYPE];\n    var script_attrs = bk_div.children[0].attributes;\n    for (var i = 0; i < script_attrs.length; i++) {\n      toinsert[toinsert.length - 1].childNodes[1].setAttribute(script_attrs[i].name, script_attrs[i].value);\n    }\n    // store reference to server id on output_area\n    output_area._bokeh_server_id = output.metadata[EXEC_MIME_TYPE][\"server_id\"];\n  }\n}\n\n/**\n * Handle when an output is cleared or removed\n */\nfunction handle_clear_output(event, handle) {\n  var id = handle.cell.output_area._hv_plot_id;\n  var server_id = handle.cell.output_area._bokeh_server_id;\n  if (((id === undefined) || !(id in PyViz.plot_index)) && (server_id !== undefined)) { return; }\n  var comm = window.PyViz.comm_manager.get_client_comm(\"hv-extension-comm\", \"hv-extension-comm\", function () {});\n  if (server_id !== null) {\n    comm.send({event_type: 'server_delete', 'id': server_id});\n    return;\n  } else if (comm !== null) {\n    comm.send({event_type: 'delete', 'id': id});\n  }\n  delete PyViz.plot_index[id];\n  if ((window.Bokeh !== undefined) & (id in window.Bokeh.index)) {\n    var doc = window.Bokeh.index[id].model.document\n    doc.clear();\n    const i = window.Bokeh.documents.indexOf(doc);\n    if (i > -1) {\n      window.Bokeh.documents.splice(i, 1);\n    }\n  }\n}\n\n/**\n * Handle kernel restart event\n */\nfunction handle_kernel_cleanup(event, handle) {\n  delete PyViz.comms[\"hv-extension-comm\"];\n  window.PyViz.plot_index = {}\n}\n\n/**\n * Handle update_display_data messages\n */\nfunction handle_update_output(event, handle) {\n  handle_clear_output(event, {cell: {output_area: handle.output_area}})\n  handle_add_output(event, handle)\n}\n\nfunction register_renderer(events, OutputArea) {\n  function append_mime(data, metadata, element) {\n    // create a DOM node to render to\n    var toinsert = this.create_output_subarea(\n    metadata,\n    CLASS_NAME,\n    EXEC_MIME_TYPE\n    );\n    this.keyboard_manager.register_events(toinsert);\n    // Render to node\n    var props = {data: data, metadata: metadata[EXEC_MIME_TYPE]};\n    render(props, toinsert[0]);\n    element.append(toinsert);\n    return toinsert\n  }\n\n  events.on('output_added.OutputArea', handle_add_output);\n  events.on('output_updated.OutputArea', handle_update_output);\n  events.on('clear_output.CodeCell', handle_clear_output);\n  events.on('delete.Cell', handle_clear_output);\n  events.on('kernel_ready.Kernel', handle_kernel_cleanup);\n\n  OutputArea.prototype.register_mime_type(EXEC_MIME_TYPE, append_mime, {\n    safe: true,\n    index: 0\n  });\n}\n\nif (window.Jupyter !== undefined) {\n  try {\n    var events = require('base/js/events');\n    var OutputArea = require('notebook/js/outputarea').OutputArea;\n    if (OutputArea.prototype.mime_types().indexOf(EXEC_MIME_TYPE) == -1) {\n      register_renderer(events, OutputArea);\n    }\n  } catch(err) {\n  }\n}\n"
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "#export\n",
    "css = '''\n",
    "\n",
    ".bk-root .bk-btn-default.bk-active {\n",
    "    background-color: #AECCE6;\n",
    "}\n",
    "\n",
    ".bk.accordion {\n",
    "    font-size: smaller;\n",
    "}\n",
    "\n",
    ".bk-root .bk-btn {\n",
    "    padding: 6px 0px 6px 0px;\n",
    "}\n",
    "\n",
    ".bk.uniprot_options {\n",
    "    margin: 17px 6px 12px 6px;\n",
    "    background-color: white;\n",
    "    font-size: 13px;\n",
    "    font-weight: initial;\n",
    "    text-align: center;\n",
    "    height: 55px;\n",
    "    border: 1px solid #e0e0e0;\n",
    "    border-radius: 2px;\n",
    "}\n",
    "\n",
    ".bk-root .bk-btn-primary {\n",
    "    background-color: #045082;\n",
    "    font-size: 11px;\n",
    "    font-weight: 700;\n",
    "    text-transform: uppercase;\n",
    "    letter-spacing: 1.5px;\n",
    "}\n",
    "\n",
    ".bk.card-title {\n",
    "    font-size: 13px;\n",
    "    font-weight: initial;\n",
    "}\n",
    "\n",
    "h1 {\n",
    "    color: #045082;\n",
    "    font-size: 45px;\n",
    "    line-height: 0.6;\n",
    "    text-align: center;\n",
    "}\n",
    "\n",
    "h4 {\n",
    "    color: #173245;\n",
    "    font-size: 18px;\n",
    "    line-height: 0.6;\n",
    "    text-align: center;\n",
    "}\n",
    "\n",
    ".codehilite {\n",
    "    color: #173245;\n",
    "    font-size: 14px;\n",
    "    background-color: #eaeaea;\n",
    "}\n",
    "\n",
    ".codehilite .s2 {\n",
    "    color: #173245;\n",
    "    font-weight: bold;\n",
    "}\n",
    "\n",
    ".codehilite .ow {\n",
    "    color: #045082;\n",
    "    font-weight: initial;\n",
    "}\n",
    "\n",
    ".bk.opt {\n",
    "    position: relative;\n",
    "    display: block;\n",
    "    left: 75px;\n",
    "    top: 0px;\n",
    "    width: 80px;\n",
    "    height: 80px;\n",
    "}\n",
    "\n",
    ".bk.main-part {\n",
    "    background-color: #EAEAEA;\n",
    "    font-size: 17px;\n",
    "    line-height: 23px;\n",
    "    letter-spacing: 0px;\n",
    "    font-weight: 500;\n",
    "    color: #045082;\n",
    "    text-align: center;\n",
    "    position: relative !important;\n",
    "    margin-left: auto;\n",
    "    margin-right: auto;\n",
    "    width: 40%;\n",
    "}\n",
    "\n",
    ".bk.selection_box {\n",
    "    background-color: #EAEAEA;\n",
    "}\n",
    "\n",
    ".bk.alert-danger {\n",
    "    background-color: #EAEAEA;\n",
    "    color: #c72d3b;\n",
    "    border: 0px #EAEAEA solid;\n",
    "    padding: 0;\n",
    "    font-size: 12px;\n",
    "}\n",
    "\n",
    "'''\n",
    "pn.extension(raw_css=[css])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Header"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "### HEADER\n",
    "header_titel = pn.pane.Markdown(\n",
    "    '# AlphaMap', \n",
    "    width=1250,\n",
    "    css_classes=['main_header']\n",
    ")\n",
    "divider = pn.pane.HTML(\n",
    "    '<hr style=\"height: 2px; border:none; background-color: #045082; width: 1480px\">', \n",
    "    width=1500, \n",
    "    align='center'\n",
    ")\n",
    "mpi_biochem_logo = pn.pane.PNG(\n",
    "    'data/img/mpi_logo.png', \n",
    "    link_url='https://www.biochem.mpg.de/en', \n",
    "    width=60, \n",
    "    height=60,\n",
    "    align='start'\n",
    ")\n",
    "mpi_logo = pn.pane.JPG(\n",
    "    'data/img/max-planck-gesellschaft.jpg', \n",
    "    link_url='https://www.biochem.mpg.de/en',\n",
    "    height=62,\n",
    "    embed=True,\n",
    "    width=62,\n",
    "    margin=(5, 0, 0, 5),\n",
    "    css_classes=['opt']\n",
    ")\n",
    "header = pn.Row(\n",
    "    mpi_biochem_logo,\n",
    "    mpi_logo,\n",
    "    header_titel,\n",
    "    height=70\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "### WIDGETS\n",
    "\n",
    "#####################################\n",
    "# SELECTORS\n",
    "select_protein = pn.widgets.AutocompleteInput(\n",
    "    name='Select a Uniprot accession number:', \n",
    "    placeholder='Type first letters ...',\n",
    "    min_characters=1\n",
    ")\n",
    "select_organism = pn.widgets.Select(\n",
    "    name='Select an organism:', \n",
    "    value='Human', \n",
    "    options=['Human', 'Mouse', 'Arabidopsis'],\n",
    "    align='center',\n",
    "    margin=(0,0,0,7),\n",
    "    width=300\n",
    ")\n",
    "\n",
    "#####################################\n",
    "# CHECKBOXES\n",
    "is_gene = pn.widgets.Checkbox(\n",
    "    name='Search by a gene name'\n",
    ")\n",
    "select_all = pn.widgets.Checkbox(\n",
    "    name='Select all options', \n",
    "    width=150\n",
    ")\n",
    "clear_all = pn.widgets.Checkbox(\n",
    "    name='Clear all options',\n",
    "    width=150\n",
    ")\n",
    "\n",
    "#####################################\n",
    "# TEXT\n",
    "upload_file_text = pn.pane.Markdown(\n",
    "    'Upload an experimental file:',\n",
    "    margin=(10,0,-5,15)\n",
    ")\n",
    "upload_file_text_inside_card = pn.pane.Markdown(\n",
    "    'Upload an experimental file:',\n",
    "    margin=(5,0,-5,15)\n",
    ")\n",
    "#####################################\n",
    "# RAW EXPERIMENTAL DATA\n",
    "\n",
    "error_message = \"The selected file can't be uploaded. Please check the instructions for data uploading.\"\n",
    "\n",
    "# first experimental file\n",
    "experimental_data = pn.widgets.FileInput(\n",
    "    accept=\".csv, .txt, .tsv\", \n",
    "    width=400,\n",
    "    margin=(-5,-30,5,15)\n",
    ")\n",
    "experimental_data_sample = pn.widgets.MultiSelect(\n",
    "    name='Select samples:', \n",
    "    disabled=True,\n",
    "    size=5,\n",
    "    width=620,\n",
    "    margin=(0,0,20,14)\n",
    ")\n",
    "experimental_data_warning = pn.pane.Alert(\n",
    "    width=550,\n",
    "    height=30,\n",
    "    alert_type=\"danger\", \n",
    "    margin=(-20,0,15,15)\n",
    ")\n",
    "experimental_data_sample_name = pn.widgets.TextInput(\n",
    "    name='Rename samples:', \n",
    "    disabled=True,\n",
    "    width=240,\n",
    "    margin=(23,0,5,10)\n",
    ")\n",
    "# second experimental file\n",
    "experimental_data_2 = pn.widgets.FileInput(\n",
    "    accept=\".csv, .txt, .tsv\", \n",
    "    width=400,\n",
    "    margin=(-5,-30,5,15)\n",
    ")\n",
    "experimental_data_2_sample = pn.widgets.MultiSelect(\n",
    "    name='Select samples:', \n",
    "    size=5,\n",
    "    disabled=True,\n",
    "    width=620,\n",
    "    margin=(0,0,0,14)\n",
    ")\n",
    "experimental_data_2_warning = pn.pane.Alert(\n",
    "    width=550,\n",
    "    height=30,\n",
    "    alert_type=\"danger\", \n",
    "    margin=(-20,0,15,15)\n",
    ")\n",
    "experimental_data_2_sample_name = pn.widgets.TextInput(\n",
    "    name='Rename samples:', \n",
    "    disabled=True,\n",
    "    width=240,\n",
    "    margin=(18,0,5,10)\n",
    ")\n",
    "# third experimental file\n",
    "experimental_data_3 = pn.widgets.FileInput(\n",
    "    accept=\".csv, .txt, .tsv\", \n",
    "    width=400,\n",
    "    margin=(-5,-30,5,15)\n",
    ")\n",
    "experimental_data_3_sample = pn.widgets.MultiSelect(\n",
    "    name='Select samples:', \n",
    "    size=5,\n",
    "    disabled=True,\n",
    "    width=620,\n",
    "    margin=(0,0,20,14)\n",
    ")\n",
    "experimental_data_3_warning = pn.pane.Alert(\n",
    "    width=550,\n",
    "    height=30,\n",
    "    alert_type=\"danger\", \n",
    "    margin=(-20,0,15,15)\n",
    ")\n",
    "experimental_data_3_sample_name = pn.widgets.TextInput(\n",
    "    name='Rename samples:', \n",
    "    disabled=True,\n",
    "    width=240,\n",
    "    margin=(18,0,5,10)\n",
    ")\n",
    "\n",
    "#####################################\n",
    "# PREPROCESSED EXPERIMENTAL DATA\n",
    "preprocessed_exp_data = pn.widgets.DataFrame(\n",
    "    name='Exp_data'\n",
    ")\n",
    "preprocessed_exp_data_2 = pn.widgets.DataFrame(\n",
    "    name='Exp_data_2'\n",
    ")\n",
    "preprocessed_exp_data_3 = pn.widgets.DataFrame(\n",
    "    name='Exp_data_3'\n",
    ")\n",
    "\n",
    "#####################################\n",
    "# BUTTONS\n",
    "upload_button = pn.widgets.Button(\n",
    "    name='Upload  Data', \n",
    "    button_type='primary', \n",
    "    css_classes=['button_options'],\n",
    "    height=40,\n",
    "    width=150,\n",
    "    align='center',\n",
    "    margin=(0,0,0,0)\n",
    ")\n",
    "upload_spinner = pn.indicators.LoadingSpinner(\n",
    "    value=False,\n",
    "    bgcolor='light',\n",
    "    color='secondary',\n",
    "    align='center',\n",
    "    margin=(0,0,20,20),\n",
    "    width=40, \n",
    "    height=40\n",
    ")\n",
    "visualize_button = pn.widgets.Button(\n",
    "    name='Visualize  Data', \n",
    "    button_type='primary', \n",
    "    css_classes=['button_options'],\n",
    "    height=40,\n",
    "    width=150,\n",
    "    align='center',\n",
    "    margin=(0,0,0,0)\n",
    ")\n",
    "visualize_spinner = pn.indicators.LoadingSpinner(\n",
    "    value=False,\n",
    "    bgcolor='light',\n",
    "    color='secondary',\n",
    "    align='center',\n",
    "    margin=(0,0,20,20),\n",
    "    width=40, \n",
    "    height=40\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "### UNIPROT OPTIONS\n",
    "options_preprocessing_events = pn.widgets.CheckButtonGroup(\n",
    "    name='Molecule processing', \n",
    "    value=['Chain', 'Initiator methionine', 'Peptide', 'Propeptide', 'Signal peptide', 'Transit peptide'], \n",
    "    options=['Chain', 'Initiator methionine', 'Peptide', 'Propeptide', 'Signal peptide', 'Transit peptide'],\n",
    "    align='center'\n",
    ")\n",
    "options_PTMs = pn.widgets.CheckButtonGroup(\n",
    "    name='Post-translational modification', \n",
    "    options=['Cross-link', 'Disulfide bond', 'Glycosylation', 'Lipidation', 'Modified residue'],\n",
    "    value=['Cross-link', 'Disulfide bond', 'Glycosylation', 'Lipidation', 'Modified residue'],\n",
    "    align='center'\n",
    ")\n",
    "options_domains = pn.widgets.CheckButtonGroup(\n",
    "    name='Family & Domain',  \n",
    "    options=['Coiled coil', 'Compositional bias', 'Domain', 'Motif', 'Region', 'Repeat', 'Zinc finger'],\n",
    "    value=['Coiled coil', 'Compositional bias', 'Domain', 'Motif', 'Region', 'Repeat', 'Zinc finger'],\n",
    "    align='center'\n",
    ")\n",
    "options_locations = pn.widgets.CheckButtonGroup(\n",
    "    name='Subcellular location', \n",
    "    options=['Intramembrane', 'Topological domain', 'Transmembrane'],\n",
    "    value=['Intramembrane', 'Topological domain', 'Transmembrane'],\n",
    "    align='center'\n",
    ")\n",
    "options_functions = pn.widgets.CheckButtonGroup(\n",
    "    name='Function', \n",
    "    options=['Active site', 'Binding site', 'Calcium binding', 'DNA binding', 'Metal binding', 'Nucleotide binding', 'Site'],\n",
    "    value=['Active site', 'Binding site', 'Calcium binding', 'DNA binding', 'Metal binding', 'Nucleotide binding', 'Site'],\n",
    "    align='center'\n",
    ")\n",
    "options_sequences = pn.widgets.CheckButtonGroup(\n",
    "    name='Sequence', \n",
    "    options=['Alternative sequence', 'Natural variant', 'Non-adjacent residues', 'Non-standard residue', \n",
    "             'Non-terminal residue', 'Sequence conflict', 'Sequence uncertainty'],\n",
    "    value=['Alternative sequence', 'Natural variant', 'Non-adjacent residues', 'Non-standard residue', \n",
    "             'Non-terminal residue', 'Sequence conflict', 'Sequence uncertainty'],\n",
    "    align='center'\n",
    ")\n",
    "options_other = pn.widgets.CheckButtonGroup(\n",
    "    name='Other options', \n",
    "    options=['Secondary structure', 'Mutagenesis'],\n",
    "    value=['Secondary structure', 'Mutagenesis'],\n",
    "    align='center'\n",
    ")\n",
    "\n",
    "#####################################\n",
    "### LAYOUTS\n",
    "uniprot_options = pn.Accordion(\n",
    "    options_preprocessing_events, \n",
    "    options_PTMs,\n",
    "    options_domains,\n",
    "    options_locations,\n",
    "    options_functions,\n",
    "    options_sequences,\n",
    "    options_other,\n",
    "    active=list(range(0,7)),\n",
    "    header_background='EAEAEA',\n",
    "    active_header_background='EAEAEA',\n",
    "    width = 850\n",
    ")\n",
    "\n",
    "uniprot_options_tab = pn.Card(\n",
    "    uniprot_options,\n",
    "    pn.Row(\n",
    "        select_all,\n",
    "        clear_all, \n",
    "        margin=5\n",
    "    ),\n",
    "    title='Uniprot Options',\n",
    "    collapsed=True,\n",
    "    header_background='EAEAEA',\n",
    "    active_header_background='EAEAEA',\n",
    "    width=860,\n",
    "    css_classes=['uniprot_options'],\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "### MAIN PART\n",
    "project_description = pn.pane.Markdown(\n",
    "    \"\"\"### This website enables the exploration of proteomic datasets on the peptide level. You can evaluate the sequence coverage of any selected protein and its identified post-translational modifications (PTMs). UniProt information is available to directly annotate sequence regions of interest such as protein domains, secondary structures, sequence variants, known PTMs, etc. \"\"\",\n",
    "    margin=(10, 0, -20, 0), \n",
    "    css_classes=['main-part'], \n",
    "    width=615\n",
    ")\n",
    "\n",
    "divider_descr = pn.pane.HTML(\n",
    "    '<hr style=\"height: 8px; border:none; background-color: #045082; width: 620px\">', \n",
    "    width=1510, \n",
    "    align='center'\n",
    ")\n",
    "\n",
    "project_instuction = pn.pane.Markdown(\n",
    "    \"\"\"#### How to use AlphaMap:\n",
    "    1. Select the organism of your proteomic study.  \n",
    "    2. Select up to 3 different bottom-up proteomic datasets analyzed  \n",
    "    by AlphaPept, MaxQuant or Spectronaut.  \n",
    "    3. Press the \"Upload Data\" button.  \n",
    "    4. Select a protein of interest.  \n",
    "    5. Select desired UniProt sequence annotations.  \n",
    "    6. Press the \"Visualize Data\" button.  \n",
    "    7. Enjoy exploring your data!  \n",
    "    \"\"\",\n",
    "    width=500,\n",
    "    align='start',\n",
    "    margin=(25, 80, 0, 10)\n",
    ")\n",
    "\n",
    "spectronaut_description = pn.pane.Markdown(\n",
    "    \"\"\"\n",
    "    The data needs to be exported in the **normal long** format as .tsv or .csv file. \n",
    "    \n",
    "    It needs to include the following columns:  \n",
    "    >- PEP.AllOccurringProteinAccessions  \n",
    "    >- EG.ModifiedSequence  \n",
    "    >- R.FileName  \n",
    "    \n",
    "    To ensure the correct export format from Spectronaut, you can download and  \n",
    "    apply the provided export scheme “spectronaut_export_scheme.rs”.\n",
    "    \"\"\",\n",
    "    width=500,\n",
    "    align='start',\n",
    "    margin=(0, 80, 0, 20)\n",
    ")\n",
    "\n",
    "spectronaut_scheme = pn.widgets.FileDownload(\n",
    "    file='data/spectronaut_export_scheme.rs',\n",
    "    filename='spectronaut_export_scheme.rs',\n",
    "    button_type='default', \n",
    "    auto=True,\n",
    "    css_classes=['button_options'],\n",
    ")\n",
    "\n",
    "maxquant_description = pn.pane.Markdown(\n",
    "    \"\"\"\n",
    "    To visualize the proteins which were analyzed by the MaxQuant software please use the **evidence.txt** file.\n",
    "\n",
    "    The following columns from the file are used for visualization:\n",
    "    >- Proteins\n",
    "    >- Modified sequence\n",
    "    >- Raw file\n",
    "    \"\"\",\n",
    "    width=500,\n",
    "    align='start',\n",
    "    margin=(0, 80, 0, 20)\n",
    ")\n",
    "\n",
    "spectronaut_instructions = pn.Card(\n",
    "    spectronaut_description,\n",
    "    spectronaut_scheme,\n",
    "    title='Spectronaut instructions',\n",
    "    collapsed=True,\n",
    "    width=505,\n",
    "    align='start',\n",
    "    margin=(10, 80, 5, 10),\n",
    "    css_classes=['spectronaut_instr']\n",
    ")\n",
    "\n",
    "maxquant_instructions = pn.Card(\n",
    "    maxquant_description,\n",
    "    title='MaxQuant instructions',\n",
    "    collapsed=True,\n",
    "    width=505,\n",
    "    align='start',\n",
    "    margin=(0, 80, 5, 10),\n",
    "    css_classes=['spectronaut_instr']\n",
    ")\n",
    "\n",
    "additional_data_card = pn.Card(\n",
    "    pn.Row(\n",
    "        pn.Column(\n",
    "            upload_file_text_inside_card,\n",
    "            experimental_data_2\n",
    "        ),\n",
    "        experimental_data_2_sample_name\n",
    "    ),\n",
    "    experimental_data_2_warning,\n",
    "    experimental_data_2_sample,\n",
    "    pn.Row(\n",
    "        pn.Column(\n",
    "            upload_file_text_inside_card,\n",
    "            experimental_data_3\n",
    "        ),\n",
    "        experimental_data_3_sample_name\n",
    "    ),\n",
    "    experimental_data_3_warning,\n",
    "    experimental_data_3_sample,\n",
    "    title='Upload additional experimental files',\n",
    "    collapsed=True,\n",
    "    width=640,\n",
    "    margin=(2,0,0,5),\n",
    "    css_classes=['add_experim_options']\n",
    ")\n",
    "\n",
    "selection_box = pn.Column(\n",
    "    select_organism,\n",
    "    pn.Row(\n",
    "        pn.Column(\n",
    "            upload_file_text,\n",
    "            experimental_data\n",
    "        ),\n",
    "        experimental_data_sample_name\n",
    "    ),\n",
    "    experimental_data_warning,\n",
    "    experimental_data_sample,\n",
    "    additional_data_card,\n",
    "    margin=(20, 150, 10, 80),\n",
    "    width=550,\n",
    "    css_classes=['selection_box'],\n",
    ")\n",
    "\n",
    "main_part = pn.Column(\n",
    "    project_description,\n",
    "    divider_descr,\n",
    "    pn.Row(\n",
    "        pn.Column(\n",
    "            project_instuction,\n",
    "            spectronaut_instructions, \n",
    "            maxquant_instructions\n",
    "        ),\n",
    "        selection_box,\n",
    "        align='center'\n",
    "    ),\n",
    "    pn.Row(\n",
    "        upload_button,\n",
    "        upload_spinner,\n",
    "        align='center'\n",
    "    ),\n",
    "    background='#eaeaea',\n",
    "    width=1510,\n",
    "    margin=(5, 0, 10, 0)\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "# switch to different websites\n",
    "uniprot_link = pn.pane.PNG(\n",
    "    'data/img/uniprot_logo.png', \n",
    "    width=120, \n",
    "    height=60,\n",
    "    align='start',\n",
    "    margin=(0, 30, 0, 40)\n",
    ")\n",
    "phosposite_link = pn.pane.PNG(\n",
    "    'data/img/phosphosite_logo.png',  \n",
    "    width=200, \n",
    "    height=60,\n",
    "    align='start',\n",
    "    margin=(0, 20)\n",
    ")\n",
    "protter_link = pn.pane.PNG(\n",
    "    'data/img/protter_logo.png', \n",
    "    width=140, \n",
    "    height=60,\n",
    "    align='start',\n",
    "    margin=(0, 20)\n",
    ")\n",
    "pdb_link = pn.pane.PNG(\n",
    "    'data/img/pdb_logo.png', \n",
    "    width=120, \n",
    "    height=60,\n",
    "    align='start',\n",
    "    margin=(0, 20)\n",
    ")\n",
    "\n",
    "def extract_uniprot_ai(protein, is_gene):\n",
    "    if is_gene:\n",
    "        return re.findall(r\"\\((?P<id>.+?)\\)\", protein)[0]\n",
    "    return protein\n",
    "        \n",
    "\n",
    "def update_all_links():\n",
    "    selected_protein_id = extract_uniprot_ai(select_protein.value, is_gene.value)\n",
    "    uniprot_link.link_url = 'https://www.uniprot.org/uniprot/' + selected_protein_id\n",
    "    phosposite_link.link_url = 'http://www.phosphosite.org/uniprotAccAction?id=' + selected_protein_id\n",
    "    protter_link.link_url = 'https://wlab.ethz.ch/protter/#up=' + selected_protein_id\n",
    "    pdb_link.link_url = f'https://www.rcsb.org/search?request=%7B%22query%22%3A%7B%22parameters%22%3A%7B%22value%22%3A%22{selected_protein_id}%22%7D%2C%22type%22%3A%22terminal%22%2C%22service%22%3A%22text%22%2C%22node_id%22%3A0%7D%2C%22return_type%22%3A%22entry%22%2C%22request_options%22%3A%7B%22pager%22%3A%7B%22start%22%3A0%2C%22rows%22%3A100%7D%2C%22scoring_strategy%22%3A%22combined%22%2C%22sort%22%3A%5B%7B%22sort_by%22%3A%22score%22%2C%22direction%22%3A%22desc%22%7D%5D%7D%2C%22request_info%22%3A%7B%22src%22%3A%22ui%22%2C%22query_id%22%3A%223407f72e3370cd10196490437be3ec87%22%7D%7D'\n",
    "    \n",
    "\n",
    "@pn.depends(visualize_button.param.clicks)\n",
    "def visualize_buttons(_):\n",
    "    if select_protein.value:\n",
    "        update_all_links()\n",
    "        buttons_layout = pn.Row(\n",
    "            pn.pane.Markdown(\n",
    "                \"### Switch to the following websites:\", \n",
    "                margin=(16, 0, 0, 0)\n",
    "            ),\n",
    "            uniprot_link,\n",
    "            phosposite_link,\n",
    "            protter_link,\n",
    "            pdb_link,\n",
    "            height=60,\n",
    "            align='start',\n",
    "            margin=(0, 0, 0, 210)\n",
    "        )\n",
    "        return buttons_layout\n",
    "    else:\n",
    "        return None\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "### PREPROCESSING\n",
    "def upload_experimental_data():\n",
    "    global ac_gene_conversion\n",
    "    if experimental_data_sample.value == ['All samples']:\n",
    "        data_samples = None\n",
    "    else:\n",
    "        data_samples = experimental_data_sample.value\n",
    "    preprocessed_exp_data.value = format_input_data(\n",
    "        df = import_data(\n",
    "            experimental_data.value, \n",
    "            dashboard=True, \n",
    "            verbose=False, \n",
    "            sample=data_samples\n",
    "        ), \n",
    "        fasta = full_fasta, \n",
    "        modification_exp = r'\\[.*?\\]', \n",
    "        verbose = False)\n",
    "    all_unique_proteins = preprocessed_exp_data.value.unique_protein_id.unique().tolist()\n",
    "    if experimental_data_2.filename:\n",
    "        if experimental_data_2_sample.value == ['All samples']:\n",
    "            data_2_samples = None\n",
    "        else:\n",
    "            data_2_samples = experimental_data_2_sample.value\n",
    "        try:\n",
    "            preprocessed_exp_data_2.value = format_input_data(\n",
    "                df = import_data(\n",
    "                    experimental_data_2.value, \n",
    "                    dashboard=True, \n",
    "                    verbose=False,\n",
    "                    sample=data_2_samples\n",
    "                ), \n",
    "                fasta = full_fasta, \n",
    "                modification_exp = r'\\[.*?\\]', \n",
    "                verbose = False)\n",
    "            all_unique_proteins.extend(preprocessed_exp_data_2.value.unique_protein_id.unique().tolist())\n",
    "        except (TypeError, AttributeError):\n",
    "            pass\n",
    "#             print(error_message)\n",
    "    if experimental_data_3.filename:\n",
    "        if experimental_data_3_sample.value == ['All samples']:\n",
    "            data_3_samples = None\n",
    "        else:\n",
    "            data_3_samples = experimental_data_3_sample.value\n",
    "        try:\n",
    "            preprocessed_exp_data_3.value = format_input_data(\n",
    "                df = import_data(\n",
    "                    experimental_data_3.value, \n",
    "                    dashboard=True, \n",
    "                    verbose=False,\n",
    "                    sample=data_3_samples\n",
    "                ), \n",
    "                fasta = full_fasta, \n",
    "                modification_exp = r'\\[.*?\\]', \n",
    "                verbose = False)\n",
    "            all_unique_proteins.extend(preprocessed_exp_data_3.value.unique_protein_id.unique().tolist())\n",
    "        except (TypeError, AttributeError):\n",
    "#             print(error_message)\n",
    "            pass\n",
    "    ac_gene_conversion = {\n",
    "        each: f\"{full_fasta.get_by_id(each).description.get('GN')} ({full_fasta.get_by_id(each).description.get('id')})\" \\\n",
    "        for each in sorted(list(set(all_unique_proteins)))}\n",
    "    # to set a selection list of availible proteins depending whether a is_gene checkbox is checked\n",
    "    if is_gene.value:\n",
    "        select_protein.options = list(ac_gene_conversion.values())\n",
    "    else:\n",
    "        select_protein.options = list(ac_gene_conversion.keys())\n",
    "    \n",
    "    \n",
    "def upload_organism_info():\n",
    "    global full_fasta, full_uniprot\n",
    "    if select_organism.value == 'Human':\n",
    "        fasta_name = \"human.fasta\"\n",
    "        uniprot_name = \"preprocessed_uniprot_human.csv\"\n",
    "    elif select_organism.value == 'Mouse':\n",
    "        fasta_name = \"mouse.fasta\"\n",
    "        uniprot_name = \"preprocessed_uniprot_mouse.csv\"\n",
    "    elif select_organism.value == 'Arabidopsis':\n",
    "        fasta_name = \"arabidopsis_thaliana.fasta\"\n",
    "        uniprot_name = \"preprocessed_uniprot_arabidopsis.csv\"\n",
    "    full_fasta = pyteomics.fasta.IndexedUniProt(os.path.join(path_to_folder_fasta_files, fasta_name))\n",
    "    full_uniprot = pd.read_csv(os.path.join(path_to_folder_fasta_files, uniprot_name))\n",
    "    \n",
    "\n",
    "def extract_samples(data):\n",
    "    \"\"\"\n",
    "    Extract information about unique sample names that present in the raw file analyzed by MaxQuant or Spectronaut.\n",
    "    \"\"\"\n",
    "    sample_columns = ['R.FileName', 'Raw file']\n",
    "    try:\n",
    "        unique_samples = pd.read_csv(StringIO(str(data, \"utf-8\")), sep=None, engine='python', \n",
    "            usecols=lambda col: col in sample_columns).iloc[:, 0].unique().tolist()\n",
    "    except:\n",
    "        raise TypeError(\"This file can't be uploaded.\")\n",
    "    return unique_samples\n",
    "\n",
    "def extract_name(filename, sample, sample_name):\n",
    "    if sample_name:\n",
    "        name = sample_name\n",
    "    elif sample and sample != ['All samples']:\n",
    "        if isinstance(sample, list):\n",
    "            name = \";\".join(sample)\n",
    "        else:\n",
    "            name = sample\n",
    "    else:\n",
    "        name = filename.split('.')[0]\n",
    "    return name\n",
    "\n",
    "@pn.depends(\n",
    "    experimental_data.param.value, \n",
    "    watch=True\n",
    ")\n",
    "def update_data_sample_info(data1):\n",
    "    experimental_data_sample.disabled = False\n",
    "    try:\n",
    "        experimental_data_warning.object = None\n",
    "        experimental_data_sample.options = ['All samples'] + extract_samples(data1) \n",
    "        experimental_data_sample.value = ['All samples']\n",
    "    except TypeError:\n",
    "        experimental_data_warning.object = error_message\n",
    "        experimental_data_sample.disabled = True\n",
    "        experimental_data_sample_name.disabled = True\n",
    "        experimental_data_sample.options = []\n",
    "        experimental_data_sample.value = []\n",
    "        experimental_data_sample_name.value = ''\n",
    "\n",
    "@pn.depends(\n",
    "    experimental_data_2.param.value,\n",
    "    watch=True\n",
    ")\n",
    "def update_data_2_sample_info(data2):\n",
    "    experimental_data_2_sample.disabled = False\n",
    "    try:\n",
    "        experimental_data_2_warning.object = None\n",
    "        experimental_data_2_sample.options = ['All samples'] + extract_samples(data2) \n",
    "        experimental_data_2_sample.value = ['All samples']\n",
    "    except TypeError:\n",
    "        experimental_data_2_warning.object = error_message\n",
    "        experimental_data_2_sample.disabled = True\n",
    "        experimental_data_2_sample_name.disabled = True\n",
    "        experimental_data_2_sample.options = []\n",
    "        experimental_data_2_sample.value = []\n",
    "        experimental_data_2_sample_name.value = ''\n",
    "    \n",
    "@pn.depends(\n",
    "    experimental_data_3.param.value,\n",
    "    watch=True\n",
    ")\n",
    "def update_data_3_sample_info(data3):\n",
    "    experimental_data_3_sample.disabled = False\n",
    "    try:\n",
    "        experimental_data_3_warning.object = None\n",
    "        experimental_data_3_sample.options = ['All samples'] + extract_samples(data3) \n",
    "        experimental_data_3_sample.value = ['All samples']\n",
    "    except TypeError:\n",
    "        experimental_data_3_warning.object = error_message\n",
    "        experimental_data_3_sample.disabled = True\n",
    "        experimental_data_3_sample_name.disabled = True\n",
    "        experimental_data_3_sample.options = []\n",
    "        experimental_data_3_sample.value = []\n",
    "        experimental_data_3_sample_name.value = ''\n",
    "    \n",
    "    \n",
    "@pn.depends(\n",
    "    experimental_data_sample.param.value,\n",
    "    experimental_data_2_sample.param.value,\n",
    "    experimental_data_3_sample.param.value,\n",
    "    watch=True\n",
    ")\n",
    "def change_sample_name_state(data_sample, data_2_sample, data_3_sample):\n",
    "    if data_sample:\n",
    "        experimental_data_sample_name.disabled = False\n",
    "    else:\n",
    "        experimental_data_sample_name.disabled = True\n",
    "    if data_2_sample:\n",
    "        experimental_data_2_sample_name.disabled = False\n",
    "    else:\n",
    "        experimental_data_2_sample_name.disabled = True\n",
    "    if data_3_sample:\n",
    "        experimental_data_3_sample_name.disabled = False \n",
    "    else:\n",
    "        experimental_data_3_sample_name.disabled = True\n",
    "\n",
    "@pn.depends(\n",
    "    select_all.param.value,\n",
    "    clear_all.param.value,\n",
    "    watch=True\n",
    ")\n",
    "def change_uniprot_selection(select, clear):\n",
    "    if select:\n",
    "        clear_all.value = False\n",
    "        select_all.value = False\n",
    "        for each in uniprot_options.objects:\n",
    "            each.value = each.options\n",
    "    if clear:\n",
    "        clear_all.value = False\n",
    "        select_all.value = False\n",
    "        for each in uniprot_options.objects:\n",
    "            each.value = []\n",
    "        \n",
    "        \n",
    "@pn.depends(\n",
    "    is_gene.param.value, \n",
    "    watch=True\n",
    ")\n",
    "def change_autocomplete_input(is_gene_selected):\n",
    "    if any(\n",
    "        [experimental_data_sample.value, experimental_data_2_sample.value, experimental_data_3_sample.value]\n",
    "    ):\n",
    "        if is_gene_selected:\n",
    "            select_protein.options = list(ac_gene_conversion.values())\n",
    "        else:\n",
    "            select_protein.options = list(ac_gene_conversion.keys())\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "### VISUALIZATION\n",
    "@pn.depends(\n",
    "    upload_button.param.clicks\n",
    ")\n",
    "def upload_data(clicks):\n",
    "    if clicks > 0 and any(\n",
    "        [experimental_data_sample.value, experimental_data_2_sample.value, experimental_data_3_sample.value]\n",
    "    ):\n",
    "        upload_spinner.value = True\n",
    "        select_protein.value = None\n",
    "        # preload the data\n",
    "        upload_organism_info()\n",
    "        upload_experimental_data()\n",
    "        # create a layout\n",
    "        app = pn.Column(\n",
    "            pn.Row(\n",
    "                pn.layout.VSpacer(width=100),\n",
    "                pn.Column(\n",
    "                    select_protein,\n",
    "                    is_gene\n",
    "                ),\n",
    "                pn.layout.VSpacer(width=80),\n",
    "                uniprot_options_tab,\n",
    "                align='center'\n",
    "            ),\n",
    "            pn.layout.HSpacer(height=4),\n",
    "            pn.Row(\n",
    "                visualize_button,\n",
    "                visualize_spinner,\n",
    "                align='center'\n",
    "            ),\n",
    "            divider,\n",
    "            pn.Pane(\n",
    "                visualize_plot, \n",
    "                config={'toImageButtonOptions': \n",
    "                           {'format': 'svg', # one of png, svg, jpeg, webp\n",
    "                            'filename': 'custom_image',\n",
    "                            'height': 500,\n",
    "                            'width': 1500,\n",
    "                            'scale': 1 # Multiply title/legend/axis/canvas sizes by this factor\n",
    "                           }\n",
    "                       }, \n",
    "                align='center',\n",
    "                width=1500\n",
    "            ),\n",
    "            visualize_buttons,\n",
    "            margin=(20, 0)\n",
    "        )\n",
    "        upload_spinner.value = False\n",
    "        return app\n",
    "    \n",
    "    \n",
    "@pn.depends(visualize_button.param.clicks)\n",
    "def visualize_plot(_):\n",
    "    if select_protein.value:\n",
    "        visualize_spinner.value = True\n",
    "        # combine selected uniprot options in one list\n",
    "        uniprot_options_combined = sum([each.value for each in uniprot_options.objects if each.value], [])\n",
    "        # extract all experimental data and names \n",
    "        all_data = []\n",
    "        all_names = []\n",
    "        if preprocessed_exp_data.value is not None:\n",
    "            all_data.append(preprocessed_exp_data.value)\n",
    "            all_names.append(\n",
    "                extract_name(experimental_data.filename, \n",
    "                             experimental_data_sample.value, \n",
    "                             experimental_data_sample_name.value\n",
    "                )\n",
    "            )\n",
    "        if preprocessed_exp_data_2.value is not None:\n",
    "            all_data.append(preprocessed_exp_data_2.value)\n",
    "            all_names.append(\n",
    "                extract_name(experimental_data_2.filename, \n",
    "                             experimental_data_2_sample.value, \n",
    "                             experimental_data_2_sample_name.value\n",
    "                )\n",
    "            )\n",
    "        if preprocessed_exp_data_3.value is not None:\n",
    "            all_data.append(preprocessed_exp_data_3.value)\n",
    "            all_names.append(\n",
    "                extract_name(experimental_data_3.filename, \n",
    "                             experimental_data_3_sample.value, \n",
    "                             experimental_data_3_sample_name.value\n",
    "                )\n",
    "            )\n",
    "        # if only one experimental file is uploaded we need to return a string for input into plot_peptide_traces\n",
    "        if len(all_data) == 1:\n",
    "            all_data = all_data[0]\n",
    "            all_names = all_names[0]\n",
    "        if is_gene.value:\n",
    "            selected_protein = re.findall(r\"\\((?P<id>.+?)\\)\", select_protein.value)[0]\n",
    "        else:\n",
    "            selected_protein = select_protein.value\n",
    "        # create a main figure\n",
    "        fig =  plot_peptide_traces(\n",
    "            df = all_data,\n",
    "            name = all_names,\n",
    "            protein = selected_protein,\n",
    "            fasta = full_fasta, \n",
    "            uniprot = full_uniprot,\n",
    "            selected_features = [uniprot_feature_dict[each] for each in uniprot_options_combined],\n",
    "            uniprot_feature_dict = uniprot_feature_dict, \n",
    "            uniprot_color_dict = uniprot_color_dict)\n",
    "        visualize_spinner.value = False\n",
    "        return fig\n",
    "    else:\n",
    "        return None"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "### JS callbacks to control the behaviour of pn.Cards\n",
    "uniprot_options_tab.jscallback(\n",
    "    collapsed=\"\"\" \n",
    "        var $container = $(\"html,body\");\n",
    "        var $scrollTo = $('.uniprot_options');\n",
    "\n",
    "        $container.animate({scrollTop: $container.offset().top + $container.scrollTop(), scrollLeft: 0},300);\n",
    "        \"\"\", \n",
    "    args={'card': uniprot_options_tab}\n",
    ");\n",
    "\n",
    "additional_data_card.jscallback(\n",
    "    collapsed=\"\"\" \n",
    "        var $container = $(\"html,body\");\n",
    "        var $scrollTo = $('.add_experim_options');\n",
    "\n",
    "        $container.animate({scrollTop: $container.offset().top + $container.scrollTop(), scrollLeft: 0},300);\n",
    "        \"\"\", \n",
    "    args={'card': additional_data_card}\n",
    ");\n",
    "\n",
    "spectronaut_instructions.jscallback(\n",
    "    collapsed=\"\"\" \n",
    "        var $container = $(\"html,body\");\n",
    "        var $scrollTo = $('.spectronaut_instr');\n",
    "\n",
    "        $container.animate({scrollTop: $container.offset().top + $container.scrollTop(), scrollLeft: 0},300);\n",
    "        \"\"\", \n",
    "    args={'card': spectronaut_instructions}\n",
    ");\n",
    "\n",
    "maxquant_instructions.jscallback(\n",
    "    collapsed=\"\"\" \n",
    "        var $container = $(\"html,body\");\n",
    "        var $scrollTo = $('.spectronaut_instr');\n",
    "\n",
    "        $container.animate({scrollTop: $container.offset().top + $container.scrollTop(), scrollLeft: 0},300);\n",
    "        \"\"\", \n",
    "    args={'card': maxquant_instructions}\n",
    ");"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "layout = pn.Column(\n",
    "    header,\n",
    "    main_part,\n",
    "    upload_data\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "layout.servable()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
