---

title: UniProt data formatting


keywords: fastai
sidebar: home_sidebar



nb_path: "nbs/Uniprot_integration.ipynb"
---
<!--

#################################################
### THIS FILE WAS AUTOGENERATED! DO NOT EDIT! ###
#################################################
# file to edit: nbs/Uniprot_integration.ipynb
# command to build the docs after a change: nbdev_build_docs

-->

<div class="container" id="notebook-container">
        
    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

</div>
    {% endraw %}

<div class="cell border-box-sizing text_cell rendered"><div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<h2 id="Integration-of-Uniprot-Data">Integration of Uniprot Data<a class="anchor-link" href="#Integration-of-Uniprot-Data"> </a></h2><p>This notebook allows to download Uniprot data and save them in a csv format.</p>
<p>The preprocessed downloaded data will include information about:</p>
<ul>
<li><strong>the known preprocessing events</strong> for proteins, such as signal peptide, transit peptide, propeptide, chain, peptide;</li>
<li>information on all available in Uniprot <strong>post translational modificatios</strong>, like modified residues (Phosphorylation, Methylation, Acetylation, etc.), Lipidation, Glycosylation, etc.;</li>
<li>information on <strong>sequence similarities</strong> with other proteins and <strong>the domain(s)</strong> present in a protein, such as domain, repeat, region, motif, etc.;</li>
<li>information on <strong>the secondary and tertiary structure</strong> of proteins, such as turn, beta strand, helix.</li>
</ul>

</div>
</div>
</div>
<div class="cell border-box-sizing text_cell rendered"><div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<h3 id="Instructions-on-how-to-download-a-UniProt-annotation-file">Instructions on how to download a UniProt annotation file<a class="anchor-link" href="#Instructions-on-how-to-download-a-UniProt-annotation-file"> </a></h3><ol>
<li>Go to the Uniprot website(<a href="https://www.uniprot.org/uniprot/">https://www.uniprot.org/uniprot/</a>), select the organism of interest in the "Popular organisms" section and click on it.</li>
<li>Click the "Download" button and select "Text" format.</li>
<li>Select the "Compressed" radio button and click "Go".</li>
<li>Unzip the downloaded file and specify the path to this file.</li>
</ol>

</div>
</div>
</div>
<div class="cell border-box-sizing text_cell rendered"><div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<h3 id="Define-helper-functions-to-preprocess-Uniprot-data">Define helper functions to preprocess Uniprot data<a class="anchor-link" href="#Define-helper-functions-to-preprocess-Uniprot-data"> </a></h3>
</div>
</div>
</div>
    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_markdown rendered_html output_subarea ">
<h4 id="extract_note" class="doc_header"><code>extract_note</code><a href="https://github.com/ibludau/alphamap/tree/master/alphamap/uniprot_integration.py#L13" class="source_link" style="float:right">[source]</a></h4><blockquote><p><code>extract_note</code>(<strong><code>string</code></strong>, <strong><code>splitted</code></strong>=<em><code>False</code></em>)</p>
</blockquote>
<p>Function to extract information about note of the protein from Uniprot using regular expression</p>

</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_markdown rendered_html output_subarea ">
<h4 id="extract_note_end" class="doc_header"><code>extract_note_end</code><a href="https://github.com/ibludau/alphamap/tree/master/alphamap/uniprot_integration.py#L24" class="source_link" style="float:right">[source]</a></h4><blockquote><p><code>extract_note_end</code>(<strong><code>string</code></strong>, <strong><code>has_mark</code></strong>=<em><code>True</code></em>)</p>
</blockquote>

</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_markdown rendered_html output_subarea ">
<h4 id="resolve_unclear_position" class="doc_header"><code>resolve_unclear_position</code><a href="https://github.com/ibludau/alphamap/tree/master/alphamap/uniprot_integration.py#L33" class="source_link" style="float:right">[source]</a></h4><blockquote><p><code>resolve_unclear_position</code>(<strong><code>value</code></strong>)</p>
</blockquote>
<p>Replace unclear position of the start/end of the modification defined as '?' with -1 and if it's defined as '?N'
or "&gt;N" - by removing the '?'/'&gt;'/'&lt;' signs</p>

</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_markdown rendered_html output_subarea ">
<h4 id="extract_positions" class="doc_header"><code>extract_positions</code><a href="https://github.com/ibludau/alphamap/tree/master/alphamap/uniprot_integration.py#L46" class="source_link" style="float:right">[source]</a></h4><blockquote><p><code>extract_positions</code>(<strong><code>posit_string</code></strong>)</p>
</blockquote>
<p>Extract isoform_id(str) and start/end positions(float) of any feature key from the string</p>

</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_markdown rendered_html output_subarea ">
<h4 id="preprocess_uniprot" class="doc_header"><code>preprocess_uniprot</code><a href="https://github.com/ibludau/alphamap/tree/master/alphamap/uniprot_integration.py#L70" class="source_link" style="float:right">[source]</a></h4><blockquote><p><code>preprocess_uniprot</code>(<strong><code>path_to_file</code></strong>)</p>
</blockquote>
<p>A complex complete function to preprocess Uniprot data from specifying the path to a flat text file
to the returning a dataframe containing information about:</p>

<pre><code>- protein_id(str)
- feature(category)
- isoform_id(str)
- start(float)
- end(float)
- note information(str)</code></pre>

</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

</div>
    {% endraw %}

<div class="cell border-box-sizing text_cell rendered"><div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<h3 id="The-following-is-a-dictionary-that-maps-feature-names-to-the-feature-entries-in-the-processed-uniprot-annotation-file">The following is a dictionary that maps feature names to the feature entries in the processed uniprot annotation file<a class="anchor-link" href="#The-following-is-a-dictionary-that-maps-feature-names-to-the-feature-entries-in-the-processed-uniprot-annotation-file"> </a></h3>
</div>
</div>
</div>
    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

</div>
    {% endraw %}

<div class="cell border-box-sizing text_cell rendered"><div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<h1 id="Export-notebook-to-script">Export notebook to script<a class="anchor-link" href="#Export-notebook-to-script"> </a></h1>
</div>
</div>
</div>
</div>
 

